.. _wrangler-app:

========
wrangler
========

Synopsis
========
.. code-block:: shell
	
	singularity run [run options...] --app wrangler <container> [app options...]

Description
===========
Run MIPWrangler on demultiplexed data.

Options
=======
.. code-block:: none
	
	# Required
	-e    A unique ID given to each sequencing run by the user.
	-l    File providing a list of samples with associated information.
	-p    Probe sets to be processed.
	-s    Sample sets to be processed.
	-x    Additional arguments passed to MIPWrangler mipSetupAndExtractByArm,
              which extracts sequences and stitches paired end reads to single
              sequences.

	# Optional
	-c    Number of available processors to use.
	-h    Print the help page.
	-k    Keep intermediate files generated by MIPWrangler.
	-m    Minimum capture length for stitching excluding probe arms.
	-n    Starting number for MIP server.
	-w    Absolute path to MIPWrangler run script.

Defaults
--------
.. code-block:: shell
	
	# Required
	-x    Default: 'none'

	# Optional
	-c    Default: 1
	-k    Default: false
	-m    Default: 100
	-n    Default: 1
	-w    Default: '/opt/bin/runMIPWranglerCurrent.sh'

Examples
========

.. code-block:: shell

	# Define variables 
	probe_sets='DR1,VAR4' 
	sample_sets='JJJ' 
	stitch_options='--stitchGapExtend=1,--overWriteDirs'

	# Run app
	singularity run \
	  -B project_resources:/opt/project_resources \
	  -B fastqs:/opt/data \
	  -B wrangler_dir:/opt/analysis \
	  --app wrangler miptools.sif \
	  -e 'example' -l 'sample_list.tsv' -p ${probe_sets} \
	  -s ${sample_sets} -x ${stitch_options}

	singularity run \
	  -B project_resources:/opt/project_resources \
	  -B fastqs:/opt/data \
	  -B wrangler_dir:/opt/analysis \
	  --app wrangler miptools.sif \
	  -c 10 -e 'example2' -l 'sample_list.tsv' -p ${probe_sets} \
	  -s ${sample_sets} -x ${stitch_options} -k
