{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# NIH SRA submission\n",
    "This notebook was used to create an NIH SRA (fastq) submission for Plasmodium falciparum MIP sequencing data in collaboration with Rosenthal Lab. It will also serve as a reference for falciparum sample submissions. \n",
    "\n",
    "### Prerequisites\n",
    "1) Prior to SRA submission, create a bioproject on https://submit.ncbi.nlm.nih.gov/subs/bioproject/  \n",
    "This is a simple procedure that produces a bioproject ID which will be associated with one or more SRA submissions.  \n",
    "2) Prior to SRA submission, submit all of the samples used in a project to NIH BioSample database.\n",
    "\n",
    "\n",
    "\n",
    "SRA submissions are made through the NIH website: https://submit.ncbi.nlm.nih.gov/subs/sra\n",
    "\n",
    "There are 5 steps in the submission portal for SRA submissions that are listed below. **The main purpose of this notebook is to create the file needed at step 4**.\n",
    "\n",
    "Step 1: Submitter details\n",
    "\n",
    "Step 2: General Info\n",
    "  * Did you already register a BioProject for this research, e.g. for the submission of the reads to SRA: **YES**\n",
    "  * Did you already register a BioSample for this sample, e.g. for the submission of the reads to SRA: **YES**\n",
    "  * Chose data release time (immediate, or future): we will chose a future date during the sumbission. Once the sumbission is finished, an email to SRA requesting removal of possible human sequences is recommended. Although our targeted sequencing should not generate any human data, and even if there are some off targets, those would not have any identifying information. Still, SRA provides this service and it makes sense to take it. Once that is done, you can publish the SRA publicly, or wait until publication date.\n",
    "  \n",
    "  \n",
    "Step 3: SRA metadata\n",
    "  * Upload a file using Excel or text format (tab-delimited): This file will be created using this notebook.\n",
    "  \n",
    "Step 4: Files\n",
    "  * This notebook will copy the fastq files to specific directories. The files will be uploaded to the FTP server of NIH at this step.\n",
    "  * The first step in the file transfer is to use command line and navigate to wherever the fastq files were saved (e.g. ssh to seekdeep server and cd to the fastq directory). Optionally, start a screen if you are comfortable using one. I'll call this location the \"data location\".\n",
    "  \n",
    "  * On the SRA submission portal choose: *FTP or Aspera Command Line file preload*\n",
    "  * click on FTP upload instructions\n",
    "  \n",
    "  * From the *data location* connect to the NIH FTP server using the credentials provided on the submission portal\n",
    "    * ftp ftp-private.ncbi.nlm.nih.gov\n",
    "    * follow instructions on the portal until step 6\n",
    "    * for step 6, instead do the following commands\n",
    "    * prompt (this should turn off interactive, if it was already off, then it'll turn it on. We want this off, so if needed repeat the \"prompt\" command to turn off interactive.\n",
    "    * mput *\n",
    "  * Once the upload is complete go back to submission portal, click **Select preload folder**\n",
    "  * Select the folder containing the uploaded files.\n",
    "  * Autofinish\n",
    "  * Continue\n",
    "  \n",
    "\n",
    "Step 5: Review and submit."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Starting the notebook\n",
    "This notebook was started with the below commands:\n",
    "```bash\n",
    "base_resources=~/git/MIPTools/base_resources\n",
    "data_dir=~/processed/analysis/2020-06/ROS_200612/\n",
    "seq_data_dir=~/raw_data\n",
    "analysis_dir=~/processed/analysis/2020-06/ROS_200612/sra_submission\n",
    "container=~/shared_bin/miptools_20200728.sif\n",
    "\n",
    "mkdir -p $analysis_dir\n",
    "\n",
    "singularity run --app jupyter \\\n",
    "    -B $base_resources:/opt/resources \\\n",
    "    -B $data_dir:/opt/data \\\n",
    "    -B $seq_data_dir:/opt/data \\\n",
    "    -B $analysis_dir:/opt/analysis \\\n",
    "    $container\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Sequencing data\n",
    "It is best to have a data folder for the specific project where sequencing data, sample data etc will be stored. In this guide the location of that folder is \"/opt/data/\". In the context of singularity MIPTools container, this means that we bind the local data directory to /opt/data/ when starting the jupyter notebook (see above). Within this directory, a subdirectory \"project_data\" contains project specific files such as sample metadata, sequencing runs metadata etc.\n",
    "\n",
    "New files generated will be saved to /opt/analysis folder or the /opt/data/project_data as appropriate.\n",
    "\n",
    "/opt/work directory was bound to the root directory of raw sequencing data. So all the raw sequencing data should be available from /opt/work from within the container."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Import necessary modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Classes reloading.\n",
      "functions reloading\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "sys.path.append(\"/opt/src\")\n",
    "import os\n",
    "import subprocess\n",
    "import pandas as pd\n",
    "import scandir"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Specify directory locations for raw data and project data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_dir = \"/opt/data/project_data/\"\n",
    "raw_data_dir = \"/opt/work/\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Multiple sequencing runs\n",
    "Many projects use multiple sequencing runs. We will use adata from 6 sequencing runs for this submission.  \n",
    "\n",
    "We typically have a run ID for each sequencing run coded as the date of the run: YYMMDD. The sequence data is stored within the raw data directory in a run directory named run ID + sequencin platform such as 190312_nextseq for a sequencing run performed on March 12, 2019 on NextSeq.  \n",
    "\n",
    "There is a sample sheet associated with the run within the run directory, typically named run ID + samples.tsv, such as 190312_samples.tsv.\n",
    "\n",
    "The fastq files generated are stored in the fastq subdirectory within the run directory (190312_nextseq/fastq).\n",
    "\n",
    "We are going to generate a dictionary pointing to file locations for each run, assuming the default values are correct."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{190312: {'Run Name': '190312_nextseq',\n",
       "  'Sample Sheet': '/opt/work/190312_nextseq/190312_samples.tsv',\n",
       "  'Fastq Dir': '/opt/work/190312_nextseq/fastq'},\n",
       " 190321: {'Run Name': '190321_nextseq',\n",
       "  'Sample Sheet': '/opt/work/190321_nextseq/190321_samples.tsv',\n",
       "  'Fastq Dir': '/opt/work/190321_nextseq/fastq'},\n",
       " 190405: {'Run Name': '190405_nextseq',\n",
       "  'Sample Sheet': '/opt/work/190405_nextseq/190405_samples.tsv',\n",
       "  'Fastq Dir': '/opt/work/190405_nextseq/fastq'},\n",
       " 190517: {'Run Name': '190517_nextseq',\n",
       "  'Sample Sheet': '/opt/work/190517_nextseq/190517_samples.tsv',\n",
       "  'Fastq Dir': '/opt/work/190517_nextseq/fastq'},\n",
       " 200316: {'Run Name': '200316_nextseq',\n",
       "  'Sample Sheet': '/opt/work/200316_nextseq/200316_samples.tsv',\n",
       "  'Fastq Dir': '/opt/work/200316_nextseq/fastq'},\n",
       " 200609: {'Run Name': '200609_nextseq',\n",
       "  'Sample Sheet': '/opt/work/200609_nextseq/200609_samples.tsv',\n",
       "  'Fastq Dir': '/opt/work/200609_nextseq/fastq'}}"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "run_ids = [190312, 190321, 190405, 190517, 200316, 200609]\n",
    "seq_platform = \"nextseq\"\n",
    "run_info = {}\n",
    "for i in run_ids:\n",
    "    run_name = str(i) + \"_\" + seq_platform\n",
    "    run_info[i] = {\"Run Name\": run_name,\n",
    "                   \"Sample Sheet\": os.path.join(raw_data_dir,\n",
    "                                                run_name,\n",
    "                                                str(i) + \"_samples.tsv\"),\n",
    "                  \"Fastq Dir\": os.path.join(raw_data_dir, run_name, \"fastq\")}\n",
    "run_info"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We have used non-standard file names for 4 of the runs for reasons beyond the scope of this guide. Below are the actual file names and the first 4 are slightly different than the default values. We'll update those file paths."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_sheets_used = [\"190312_xs_samples.tsv\",\n",
    "                 \"190321_xs_samples.tsv\",\n",
    "                 \"190405_xs_samples.tsv\",\n",
    "                 \"190517_samples_xs.tsv\",\n",
    "                 \"200316_samples.tsv\",\n",
    "                 \"200609_samples.tsv\"] "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "run_info[190312][\"Sample Sheet\"] = \"/opt/work/190312_nextseq/190312_xs_samples.tsv\"\n",
    "run_info[190321][\"Sample Sheet\"] = \"/opt/work/190321_nextseq/190321_xs_samples.tsv\"\n",
    "run_info[190405][\"Sample Sheet\"] = \"/opt/work/190405_nextseq/190405_xs_samples.tsv\"\n",
    "run_info[190517][\"Sample Sheet\"] = \"/opt/work/190517_nextseq/190517_samples_xs.tsv\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### What is in a sample sheet\n",
    "When preparing sequencing libraries, we use certain terminology specific to our runs. Sample sheets provide all the information needed for connecting sequencing data to specific libraries. A \n",
    "\n",
    "  * **sample:** DNA source identifier. Multiple DNA extractions from the same source would have the same sample name, for example.\n",
    "  * **library:** A sequencing library whose data can be uniquely identified (i.e. has unique sample barcodes). Multiple libraries can be generated from the same DNA sample (replicates) during the same library prep.\n",
    "  * **Sample ID** library identifier.\n",
    "  * **Library Prep:** A library preparation identifier, typically a date. A failed library for a sample can be re-prepared in a later library prep, for example. The new library would have a new barcode pair that identifies it and distinguishes from the first one.\n",
    "\n",
    "\n",
    "There are 4 fields in the sample sheet that defines a library: **sample_name, sample_set, replicate, Library Prep**. \n",
    "  * **only alphanumeric characters and dash or underscores are allowed** in these (and most other) fields.\n",
    "  * **sample_name:** The DNA sample that the library was prepared from. Each unique DNA sample in the lab must have a unique ID. \n",
    "  * **sample_set:** This is a project specific notation to identify samples belonging to the same project. Libraries sharing this value would be normally analysed together.\n",
    "  * **replicate:** Replicate number. \n",
    "  \n",
    "  Each library needs a unique identifier, although we sometimes use the same sample multiple times in a sequencing run. The most common use is the control DNA. To create a unique Sample ID (i.e. Library ID) we combine the sample_name, sample_set and replicate fields. \n",
    "  \n",
    "  For example, we used 3d7 control DNA in 2 libraries during a library prep. sample_name: 3d7, sample_set: con, replicate: 1 and sample_name: 3d7, sample_set: con, replicate: 2. Let's say the Library Prep ID is 190201 (prepared on February 1, 2019). Corresponding Sample IDs would be 3d7-con-1 and 3d7-con-2. The sequencing data for these libraries will be permanently linked to the sample IDs (fastq file names will contain these IDs). \n",
    "  \n",
    "  Let's say we have another library prep one month later that also uses 3d7 as control. sample_name: 3d7, sample_set: con, replicate: 1, and Library Prep 190301. Sample ID: 3d7-con-1\n",
    "  \n",
    "  If we are analysing the two data sets together, we'd have 2 libraries with the same ID 3d7-con-1. In this case, we'll need to change one of these to 3d7-con-3, but we will need to keep track of what it originally was. We do this by keeping the Library Prep identifier along with the original Sample ID (Original SID field). i.e. Sample ID 3d7-con-1 = Original SID 3d7-con-1 + Library Prep 190201 and Sample ID 3d7-con-3 = Original SID 3d7-con-1 and Library Prep 190301. All downstream analysis refer to the Sample ID.\n",
    "  \n",
    "  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get fastq file information and connect with sample information\n",
    "We have the sample information for each run in the sample files provided. We also know the fastq directory for each run. Looping through the runs, we'll connect the two. \n",
    "\n",
    "We expect some samples not to have data, so when sample information and fastq information is merged, we would expect some decrease in the size of the sample tables. We are going to check that that is the case for each run. If there is an increase, on the other hand, that would point to some error that needs to be taken care of."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "190312 (3776, 6)\n",
      "(3302, 10)\n",
      "190321 (3856, 6)\n",
      "(3282, 10)\n",
      "190405 (6862, 6)\n",
      "(3030, 10)\n",
      "190517 (6240, 6)\n",
      "(664, 10)\n",
      "200316 (3146, 6)\n",
      "(3144, 10)\n",
      "200609 (1570, 6)\n",
      "(1568, 10)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Original SID</th>\n",
       "      <th>Read Order</th>\n",
       "      <th>File Name</th>\n",
       "      <th>File Path</th>\n",
       "      <th>Run Name</th>\n",
       "      <th>Run ID</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>sample_set</th>\n",
       "      <th>replicate</th>\n",
       "      <th>Library Prep</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>22801-JHU-1</td>\n",
       "      <td>1</td>\n",
       "      <td>22801-JHU-1_S1825_R1_001.fastq.gz</td>\n",
       "      <td>/opt/work/190312_nextseq/fastq/22801-JHU-1_S18...</td>\n",
       "      <td>190312_nextseq</td>\n",
       "      <td>190312</td>\n",
       "      <td>22801</td>\n",
       "      <td>JHU</td>\n",
       "      <td>1</td>\n",
       "      <td>190312</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>22801-JHU-1</td>\n",
       "      <td>2</td>\n",
       "      <td>22801-JHU-1_S1825_R2_001.fastq.gz</td>\n",
       "      <td>/opt/work/190312_nextseq/fastq/22801-JHU-1_S18...</td>\n",
       "      <td>190312_nextseq</td>\n",
       "      <td>190312</td>\n",
       "      <td>22801</td>\n",
       "      <td>JHU</td>\n",
       "      <td>1</td>\n",
       "      <td>190312</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>RXS480-ROS-1</td>\n",
       "      <td>2</td>\n",
       "      <td>RXS480-ROS-1_S953_R2_001.fastq.gz</td>\n",
       "      <td>/opt/work/190312_nextseq/fastq/RXS480-ROS-1_S9...</td>\n",
       "      <td>190312_nextseq</td>\n",
       "      <td>190312</td>\n",
       "      <td>RXS480</td>\n",
       "      <td>ROS</td>\n",
       "      <td>1</td>\n",
       "      <td>190312</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>RXS480-ROS-1</td>\n",
       "      <td>1</td>\n",
       "      <td>RXS480-ROS-1_S953_R1_001.fastq.gz</td>\n",
       "      <td>/opt/work/190312_nextseq/fastq/RXS480-ROS-1_S9...</td>\n",
       "      <td>190312_nextseq</td>\n",
       "      <td>190312</td>\n",
       "      <td>RXS480</td>\n",
       "      <td>ROS</td>\n",
       "      <td>1</td>\n",
       "      <td>190312</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>RXS514-ROS-1</td>\n",
       "      <td>1</td>\n",
       "      <td>RXS514-ROS-1_S996_R1_001.fastq.gz</td>\n",
       "      <td>/opt/work/190312_nextseq/fastq/RXS514-ROS-1_S9...</td>\n",
       "      <td>190312_nextseq</td>\n",
       "      <td>190312</td>\n",
       "      <td>RXS514</td>\n",
       "      <td>ROS</td>\n",
       "      <td>1</td>\n",
       "      <td>190312</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Original SID Read Order                          File Name  \\\n",
       "0   22801-JHU-1          1  22801-JHU-1_S1825_R1_001.fastq.gz   \n",
       "1   22801-JHU-1          2  22801-JHU-1_S1825_R2_001.fastq.gz   \n",
       "2  RXS480-ROS-1          2  RXS480-ROS-1_S953_R2_001.fastq.gz   \n",
       "3  RXS480-ROS-1          1  RXS480-ROS-1_S953_R1_001.fastq.gz   \n",
       "4  RXS514-ROS-1          1  RXS514-ROS-1_S996_R1_001.fastq.gz   \n",
       "\n",
       "                                           File Path        Run Name  Run ID  \\\n",
       "0  /opt/work/190312_nextseq/fastq/22801-JHU-1_S18...  190312_nextseq  190312   \n",
       "1  /opt/work/190312_nextseq/fastq/22801-JHU-1_S18...  190312_nextseq  190312   \n",
       "2  /opt/work/190312_nextseq/fastq/RXS480-ROS-1_S9...  190312_nextseq  190312   \n",
       "3  /opt/work/190312_nextseq/fastq/RXS480-ROS-1_S9...  190312_nextseq  190312   \n",
       "4  /opt/work/190312_nextseq/fastq/RXS514-ROS-1_S9...  190312_nextseq  190312   \n",
       "\n",
       "  sample_name sample_set  replicate  Library Prep  \n",
       "0       22801        JHU          1        190312  \n",
       "1       22801        JHU          1        190312  \n",
       "2      RXS480        ROS          1        190312  \n",
       "3      RXS480        ROS          1        190312  \n",
       "4      RXS514        ROS          1        190312  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fastq_df_list = []\n",
    "for r in run_info:\n",
    "    fastq_list = []\n",
    "    run_name = run_info[r][\"Run Name\"]\n",
    "    fastq_dir = run_info[r][\"Fastq Dir\"]\n",
    "    for entry in scandir.scandir(fastq_dir):\n",
    "        try:\n",
    "            split_entry = entry.name.split(\"_\")\n",
    "            fastq_list.append([split_entry[0],\n",
    "                               split_entry[2][-1],\n",
    "                               entry.name,\n",
    "                               entry.path,\n",
    "                               run_name,\n",
    "                               r])\n",
    "        except IndexError:\n",
    "            continue\n",
    "    fdf = pd.DataFrame(fastq_list, columns = [\"Original SID\",\n",
    "                                   \"Read Order\",\n",
    "                                   \"File Name\",\n",
    "                                   \"File Path\", \n",
    "                                   \"Run Name\", \n",
    "                                   \"Run ID\"])\n",
    "    print(r, fdf.shape)\n",
    "    sheet_file = run_info[r][\"Sample Sheet\"]\n",
    "    sheet_df = pd.read_table(sheet_file)[[\"sample_name\", \"sample_set\", \"replicate\", \"Library Prep\"]]\n",
    "    sheet_df[\"Original SID\"] = sheet_df[[\"sample_name\", \"sample_set\", \"replicate\"]].apply(\n",
    "        lambda a: \"-\".join(map(str, a)), axis=1)\n",
    "    fdf = fdf.merge(sheet_df)\n",
    "    print(fdf.shape)\n",
    "    fastq_df_list.append(fdf)\n",
    "fastq_df = pd.concat(fastq_df_list, ignore_index=True)\n",
    "fastq_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we have a table that tracks sample_name, sample_set, replicate, Library Prep fields that identify the unique library and the locations of fastq files from each sequencing run."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Select project specific libraries\n",
    "Most MIP sequencing runs have libraries belonging to different projects. We will select those that belong to this project. \n",
    "\n",
    "First, load the sample files provided and visualize the unique sample sets."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>384 Column</th>\n",
       "      <th>FW_plate</th>\n",
       "      <th>Library Prep</th>\n",
       "      <th>REV_plate</th>\n",
       "      <th>Run Name</th>\n",
       "      <th>capture_plate</th>\n",
       "      <th>capture_plate_column</th>\n",
       "      <th>capture_plate_row</th>\n",
       "      <th>diff</th>\n",
       "      <th>fw</th>\n",
       "      <th>owner</th>\n",
       "      <th>quadrant</th>\n",
       "      <th>replicate</th>\n",
       "      <th>rev</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>sample_plate</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_set</th>\n",
       "      <th>probe_set</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AG</th>\n",
       "      <th>HeOME_subset</th>\n",
       "      <td>Odd</td>\n",
       "      <td>FW-25</td>\n",
       "      <td>200316</td>\n",
       "      <td>REV-1</td>\n",
       "      <td>200316_nextseq</td>\n",
       "      <td>200305_HeOME_Begoro_P1</td>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>193</td>\n",
       "      <td>jonathan</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>07GHR5001</td>\n",
       "      <td>2014_Begoro_Hospital_Plate3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HRP</th>\n",
       "      <th>HRP2,HRP3,HRPF</th>\n",
       "      <td>Odd</td>\n",
       "      <td>FW-18</td>\n",
       "      <td>190312</td>\n",
       "      <td>REV-2</td>\n",
       "      <td>190312_nextseq</td>\n",
       "      <td>DR2_HRP_ROS10_JP_Pilot</td>\n",
       "      <td>1</td>\n",
       "      <td>E</td>\n",
       "      <td>-5.0</td>\n",
       "      <td>102</td>\n",
       "      <td>patrick</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1</td>\n",
       "      <td>97</td>\n",
       "      <td>4624-T</td>\n",
       "      <td>2018 EPHI WHO Plate 1 (Pilot)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>JHU</th>\n",
       "      <th>IBC</th>\n",
       "      <td>Odd</td>\n",
       "      <td>FW-5</td>\n",
       "      <td>180723</td>\n",
       "      <td>REV-1</td>\n",
       "      <td>190312_nextseq</td>\n",
       "      <td>IBC_Giovanna_Zambia_Plate_1</td>\n",
       "      <td>11</td>\n",
       "      <td>B</td>\n",
       "      <td>360.0</td>\n",
       "      <td>23</td>\n",
       "      <td>patrick</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>383</td>\n",
       "      <td>22755</td>\n",
       "      <td>IBC_JHU_Recapture_P1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">JJJ</th>\n",
       "      <th>DR2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>190312</td>\n",
       "      <td>NaN</td>\n",
       "      <td>190312_nextseq</td>\n",
       "      <td>190117_DR2_MIPfest_2019_P01</td>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>-12.0</td>\n",
       "      <td>193</td>\n",
       "      <td>Patrick</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>181</td>\n",
       "      <td>FP20443</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IBC</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>190312</td>\n",
       "      <td>NaN</td>\n",
       "      <td>190312_nextseq</td>\n",
       "      <td>IBC_DR2_MIPfest_2019_P05</td>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>-264.0</td>\n",
       "      <td>289</td>\n",
       "      <td>Maddi</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>25</td>\n",
       "      <td>FP20443</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">OPT</th>\n",
       "      <th>DR2</th>\n",
       "      <td>Odd</td>\n",
       "      <td>FW-12</td>\n",
       "      <td>200316</td>\n",
       "      <td>REV-3</td>\n",
       "      <td>200316_nextseq</td>\n",
       "      <td>191125_VeriFi_polymerase_Test</td>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>12</td>\n",
       "      <td>deborah</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>193</td>\n",
       "      <td>3D7-1</td>\n",
       "      <td>CTL-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HeOME_subset</th>\n",
       "      <td>Odd</td>\n",
       "      <td>FW-74</td>\n",
       "      <td>200316</td>\n",
       "      <td>REV-2</td>\n",
       "      <td>200316_nextseq</td>\n",
       "      <td>200123_HeOMEpanel</td>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>349</td>\n",
       "      <td>deborah</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>97</td>\n",
       "      <td>3D7</td>\n",
       "      <td>CTL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ROS</th>\n",
       "      <th>DR2</th>\n",
       "      <td>Odd</td>\n",
       "      <td>FW-30</td>\n",
       "      <td>190312</td>\n",
       "      <td>REV-3</td>\n",
       "      <td>190312_nextseq</td>\n",
       "      <td>DR2_HRP_ROS10_JP_Pilot</td>\n",
       "      <td>1</td>\n",
       "      <td>I</td>\n",
       "      <td>-5.0</td>\n",
       "      <td>198</td>\n",
       "      <td>patrick</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1</td>\n",
       "      <td>193</td>\n",
       "      <td>HC</td>\n",
       "      <td>CTL_DR2_HRP_ROS10_JP_Pilot1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>uganda</th>\n",
       "      <th>DR2</th>\n",
       "      <td>Odd</td>\n",
       "      <td>FW-3</td>\n",
       "      <td>200316</td>\n",
       "      <td>REV-1</td>\n",
       "      <td>200316_nextseq</td>\n",
       "      <td>Prism_recapture_plate1</td>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "      <td>deborah</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>TO-04-01</td>\n",
       "      <td>plate_1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          384 Column FW_plate  Library Prep REV_plate  \\\n",
       "sample_set probe_set                                                    \n",
       "AG         HeOME_subset          Odd    FW-25        200316     REV-1   \n",
       "HRP        HRP2,HRP3,HRPF        Odd    FW-18        190312     REV-2   \n",
       "JHU        IBC                   Odd     FW-5        180723     REV-1   \n",
       "JJJ        DR2                   NaN      NaN        190312       NaN   \n",
       "           IBC                   NaN      NaN        190312       NaN   \n",
       "OPT        DR2                   Odd    FW-12        200316     REV-3   \n",
       "           HeOME_subset          Odd    FW-74        200316     REV-2   \n",
       "ROS        DR2                   Odd    FW-30        190312     REV-3   \n",
       "uganda     DR2                   Odd     FW-3        200316     REV-1   \n",
       "\n",
       "                                 Run Name                  capture_plate  \\\n",
       "sample_set probe_set                                                       \n",
       "AG         HeOME_subset    200316_nextseq         200305_HeOME_Begoro_P1   \n",
       "HRP        HRP2,HRP3,HRPF  190312_nextseq         DR2_HRP_ROS10_JP_Pilot   \n",
       "JHU        IBC             190312_nextseq    IBC_Giovanna_Zambia_Plate_1   \n",
       "JJJ        DR2             190312_nextseq    190117_DR2_MIPfest_2019_P01   \n",
       "           IBC             190312_nextseq       IBC_DR2_MIPfest_2019_P05   \n",
       "OPT        DR2             200316_nextseq  191125_VeriFi_polymerase_Test   \n",
       "           HeOME_subset    200316_nextseq              200123_HeOMEpanel   \n",
       "ROS        DR2             190312_nextseq         DR2_HRP_ROS10_JP_Pilot   \n",
       "uganda     DR2             200316_nextseq         Prism_recapture_plate1   \n",
       "\n",
       "                           capture_plate_column capture_plate_row   diff   fw  \\\n",
       "sample_set probe_set                                                            \n",
       "AG         HeOME_subset                       1                 A    NaN  193   \n",
       "HRP        HRP2,HRP3,HRPF                     1                 E   -5.0  102   \n",
       "JHU        IBC                               11                 B  360.0   23   \n",
       "JJJ        DR2                                1                 A  -12.0  193   \n",
       "           IBC                                1                 A -264.0  289   \n",
       "OPT        DR2                                1                 A    NaN   12   \n",
       "           HeOME_subset                       1                 A    NaN  349   \n",
       "ROS        DR2                                1                 I   -5.0  198   \n",
       "uganda     DR2                                1                 A    NaN    3   \n",
       "\n",
       "                              owner  quadrant  replicate  rev sample_name  \\\n",
       "sample_set probe_set                                                        \n",
       "AG         HeOME_subset    jonathan       1.0          1    1   07GHR5001   \n",
       "HRP        HRP2,HRP3,HRPF   patrick       2.0          1   97      4624-T   \n",
       "JHU        IBC              patrick       1.0          1  383       22755   \n",
       "JJJ        DR2              Patrick       NaN          1  181     FP20443   \n",
       "           IBC                Maddi       NaN          2   25     FP20443   \n",
       "OPT        DR2              deborah       1.0          1  193       3D7-1   \n",
       "           HeOME_subset     deborah       1.0          1   97         3D7   \n",
       "ROS        DR2              patrick       3.0          1  193          HC   \n",
       "uganda     DR2              deborah       1.0          1    1    TO-04-01   \n",
       "\n",
       "                                            sample_plate  \n",
       "sample_set probe_set                                      \n",
       "AG         HeOME_subset      2014_Begoro_Hospital_Plate3  \n",
       "HRP        HRP2,HRP3,HRPF  2018 EPHI WHO Plate 1 (Pilot)  \n",
       "JHU        IBC                      IBC_JHU_Recapture_P1  \n",
       "JJJ        DR2                                       NaN  \n",
       "           IBC                                       NaN  \n",
       "OPT        DR2                                     CTL-1  \n",
       "           HeOME_subset                              CTL  \n",
       "ROS        DR2               CTL_DR2_HRP_ROS10_JP_Pilot1  \n",
       "uganda     DR2                                   plate_1  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_sheet_list = []\n",
    "for r in run_info:\n",
    "    s_sheet = pd.read_table(run_info[r][\"Sample Sheet\"])\n",
    "    s_sheet[\"Run Name\"] = run_info[r][\"Run Name\"]\n",
    "    sample_sheet_list.append(s_sheet)\n",
    "sample_sheets = pd.concat(sample_sheet_list, axis=0, ignore_index=True)\n",
    "sample_sheets.groupby([\"sample_set\", \"probe_set\"]).first()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Only ROS and uganda sets are relevant for this submission. So we'll limit the sample table to those."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(3966, 18)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>384 Column</th>\n",
       "      <th>FW_plate</th>\n",
       "      <th>Library Prep</th>\n",
       "      <th>REV_plate</th>\n",
       "      <th>Run Name</th>\n",
       "      <th>capture_plate</th>\n",
       "      <th>capture_plate_column</th>\n",
       "      <th>capture_plate_row</th>\n",
       "      <th>diff</th>\n",
       "      <th>fw</th>\n",
       "      <th>owner</th>\n",
       "      <th>probe_set</th>\n",
       "      <th>quadrant</th>\n",
       "      <th>replicate</th>\n",
       "      <th>rev</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>sample_plate</th>\n",
       "      <th>sample_set</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>366</th>\n",
       "      <td>Odd</td>\n",
       "      <td>FW-30</td>\n",
       "      <td>190312</td>\n",
       "      <td>REV-3</td>\n",
       "      <td>190312_nextseq</td>\n",
       "      <td>DR2_HRP_ROS10_JP_Pilot</td>\n",
       "      <td>1</td>\n",
       "      <td>I</td>\n",
       "      <td>-5.0</td>\n",
       "      <td>198</td>\n",
       "      <td>patrick</td>\n",
       "      <td>DR2</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1</td>\n",
       "      <td>193</td>\n",
       "      <td>HC</td>\n",
       "      <td>CTL_DR2_HRP_ROS10_JP_Pilot1</td>\n",
       "      <td>ROS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>368</th>\n",
       "      <td>Odd</td>\n",
       "      <td>FW-30</td>\n",
       "      <td>190312</td>\n",
       "      <td>REV-3</td>\n",
       "      <td>190312_nextseq</td>\n",
       "      <td>DR2_HRP_ROS10_JP_Pilot</td>\n",
       "      <td>3</td>\n",
       "      <td>I</td>\n",
       "      <td>-5.0</td>\n",
       "      <td>199</td>\n",
       "      <td>patrick</td>\n",
       "      <td>DR2</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2</td>\n",
       "      <td>194</td>\n",
       "      <td>HC</td>\n",
       "      <td>CTL_DR2_HRP_ROS10_JP_Pilot1</td>\n",
       "      <td>ROS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>370</th>\n",
       "      <td>Odd</td>\n",
       "      <td>FW-30</td>\n",
       "      <td>190312</td>\n",
       "      <td>REV-3</td>\n",
       "      <td>190312_nextseq</td>\n",
       "      <td>DR2_HRP_ROS10_JP_Pilot</td>\n",
       "      <td>5</td>\n",
       "      <td>I</td>\n",
       "      <td>-5.0</td>\n",
       "      <td>200</td>\n",
       "      <td>patrick</td>\n",
       "      <td>DR2</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1</td>\n",
       "      <td>195</td>\n",
       "      <td>LC</td>\n",
       "      <td>CTL_DR2_HRP_ROS10_JP_Pilot1</td>\n",
       "      <td>ROS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>372</th>\n",
       "      <td>Odd</td>\n",
       "      <td>FW-30</td>\n",
       "      <td>190312</td>\n",
       "      <td>REV-3</td>\n",
       "      <td>190312_nextseq</td>\n",
       "      <td>DR2_HRP_ROS10_JP_Pilot</td>\n",
       "      <td>7</td>\n",
       "      <td>I</td>\n",
       "      <td>-5.0</td>\n",
       "      <td>201</td>\n",
       "      <td>patrick</td>\n",
       "      <td>DR2</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2</td>\n",
       "      <td>196</td>\n",
       "      <td>LC</td>\n",
       "      <td>CTL_DR2_HRP_ROS10_JP_Pilot1</td>\n",
       "      <td>ROS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>374</th>\n",
       "      <td>Odd</td>\n",
       "      <td>FW-30</td>\n",
       "      <td>190312</td>\n",
       "      <td>REV-3</td>\n",
       "      <td>190312_nextseq</td>\n",
       "      <td>DR2_HRP_ROS10_JP_Pilot</td>\n",
       "      <td>9</td>\n",
       "      <td>I</td>\n",
       "      <td>-5.0</td>\n",
       "      <td>202</td>\n",
       "      <td>patrick</td>\n",
       "      <td>DR2</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1</td>\n",
       "      <td>197</td>\n",
       "      <td>NTC</td>\n",
       "      <td>CTL_DR2_HRP_ROS10_JP_Pilot1</td>\n",
       "      <td>ROS</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    384 Column FW_plate  Library Prep REV_plate        Run Name  \\\n",
       "366        Odd    FW-30        190312     REV-3  190312_nextseq   \n",
       "368        Odd    FW-30        190312     REV-3  190312_nextseq   \n",
       "370        Odd    FW-30        190312     REV-3  190312_nextseq   \n",
       "372        Odd    FW-30        190312     REV-3  190312_nextseq   \n",
       "374        Odd    FW-30        190312     REV-3  190312_nextseq   \n",
       "\n",
       "              capture_plate  capture_plate_column capture_plate_row  diff  \\\n",
       "366  DR2_HRP_ROS10_JP_Pilot                     1                 I  -5.0   \n",
       "368  DR2_HRP_ROS10_JP_Pilot                     3                 I  -5.0   \n",
       "370  DR2_HRP_ROS10_JP_Pilot                     5                 I  -5.0   \n",
       "372  DR2_HRP_ROS10_JP_Pilot                     7                 I  -5.0   \n",
       "374  DR2_HRP_ROS10_JP_Pilot                     9                 I  -5.0   \n",
       "\n",
       "      fw    owner probe_set  quadrant  replicate  rev sample_name  \\\n",
       "366  198  patrick       DR2       3.0          1  193          HC   \n",
       "368  199  patrick       DR2       3.0          2  194          HC   \n",
       "370  200  patrick       DR2       3.0          1  195          LC   \n",
       "372  201  patrick       DR2       3.0          2  196          LC   \n",
       "374  202  patrick       DR2       3.0          1  197         NTC   \n",
       "\n",
       "                    sample_plate sample_set  \n",
       "366  CTL_DR2_HRP_ROS10_JP_Pilot1        ROS  \n",
       "368  CTL_DR2_HRP_ROS10_JP_Pilot1        ROS  \n",
       "370  CTL_DR2_HRP_ROS10_JP_Pilot1        ROS  \n",
       "372  CTL_DR2_HRP_ROS10_JP_Pilot1        ROS  \n",
       "374  CTL_DR2_HRP_ROS10_JP_Pilot1        ROS  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_sheets = sample_sheets.loc[(sample_sheets[\"sample_set\"].isin([\"ROS\", \"uganda\"]))]\n",
    "print(sample_sheets.shape)\n",
    "sample_sheets.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Also, drop the negative controls. These libraries should be inspected at the data analyses step for contaminations but there is no need to upload them to the SRA."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(3929, 18)"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_sheets = sample_sheets.loc[~sample_sheets[\"sample_name\"].isin([\"NTP\", \"NTC\"])]\n",
    "sample_sheets.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(14990, 10)"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fastq_df.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Make sure there are no duplicates in the sample table. If there are, it should be inspected."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(3929, 18)"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_sheets.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(3929, 18)"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_sheets.drop_duplicates().shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Merge the fastq information with the sample table using the common column names to remove the fastq entries that do not belong to this project."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "fastq_df = fastq_df.merge(sample_sheets[[\n",
    "    \"sample_name\", \"sample_set\", \"replicate\", \"Library Prep\", \"Run Name\"]])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Linking the Sample Id to the fastqs\n",
    "We know what Original Sample ID each fastq file belongs to but the downstream analysis links data to Sample ID. These can be different from the original as explained above. We'll link the final Sample ID to the fastq files using the run metadata generated during downstream analysis. This file should be placed in the data directory. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample ID</th>\n",
       "      <th>Library Prep</th>\n",
       "      <th>384 Column</th>\n",
       "      <th>FW_plate</th>\n",
       "      <th>Original SID</th>\n",
       "      <th>REV_plate</th>\n",
       "      <th>capture_plate</th>\n",
       "      <th>capture_plate_column</th>\n",
       "      <th>capture_plate_row</th>\n",
       "      <th>diff</th>\n",
       "      <th>fw</th>\n",
       "      <th>owner</th>\n",
       "      <th>probe_set</th>\n",
       "      <th>quadrant</th>\n",
       "      <th>rev</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>sample_plate</th>\n",
       "      <th>sample_set</th>\n",
       "      <th>replicate</th>\n",
       "      <th>Sample Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3D7-ROS-1</td>\n",
       "      <td>190517</td>\n",
       "      <td>Odd</td>\n",
       "      <td>FW-24</td>\n",
       "      <td>3D7-ROS-1</td>\n",
       "      <td>REV-2</td>\n",
       "      <td>DR2_ROS_RECAP_Plate_02</td>\n",
       "      <td>1</td>\n",
       "      <td>E</td>\n",
       "      <td>NaN</td>\n",
       "      <td>108</td>\n",
       "      <td>deborah</td>\n",
       "      <td>DR2</td>\n",
       "      <td>2.0</td>\n",
       "      <td>97</td>\n",
       "      <td>3D7</td>\n",
       "      <td>CTL</td>\n",
       "      <td>ROS</td>\n",
       "      <td>1</td>\n",
       "      <td>3D7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3D7-ROS-2</td>\n",
       "      <td>190517</td>\n",
       "      <td>Odd</td>\n",
       "      <td>FW-44</td>\n",
       "      <td>3D7-ROS-2</td>\n",
       "      <td>REV-4</td>\n",
       "      <td>DR2_ROS_RECAP_Plate_01</td>\n",
       "      <td>1</td>\n",
       "      <td>M</td>\n",
       "      <td>NaN</td>\n",
       "      <td>296</td>\n",
       "      <td>deborah</td>\n",
       "      <td>DR2</td>\n",
       "      <td>4.0</td>\n",
       "      <td>289</td>\n",
       "      <td>3D7</td>\n",
       "      <td>CTL</td>\n",
       "      <td>ROS</td>\n",
       "      <td>2</td>\n",
       "      <td>3D7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3D7-ROS-3</td>\n",
       "      <td>190517</td>\n",
       "      <td>Odd</td>\n",
       "      <td>FW-24</td>\n",
       "      <td>3D7-ROS-3</td>\n",
       "      <td>REV-2</td>\n",
       "      <td>DR2_ROS_RECAP_Plate_02</td>\n",
       "      <td>3</td>\n",
       "      <td>E</td>\n",
       "      <td>NaN</td>\n",
       "      <td>97</td>\n",
       "      <td>deborah</td>\n",
       "      <td>DR2</td>\n",
       "      <td>2.0</td>\n",
       "      <td>98</td>\n",
       "      <td>3D7</td>\n",
       "      <td>CTL</td>\n",
       "      <td>ROS</td>\n",
       "      <td>3</td>\n",
       "      <td>3D7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3D7-ROS-4</td>\n",
       "      <td>190517</td>\n",
       "      <td>Odd</td>\n",
       "      <td>FW-44</td>\n",
       "      <td>3D7-ROS-4</td>\n",
       "      <td>REV-4</td>\n",
       "      <td>DR2_ROS_RECAP_Plate_01</td>\n",
       "      <td>3</td>\n",
       "      <td>M</td>\n",
       "      <td>NaN</td>\n",
       "      <td>297</td>\n",
       "      <td>deborah</td>\n",
       "      <td>DR2</td>\n",
       "      <td>4.0</td>\n",
       "      <td>290</td>\n",
       "      <td>3D7</td>\n",
       "      <td>CTL</td>\n",
       "      <td>ROS</td>\n",
       "      <td>4</td>\n",
       "      <td>3D7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3D7-uganda-1</td>\n",
       "      <td>200316</td>\n",
       "      <td>Odd</td>\n",
       "      <td>FW-47</td>\n",
       "      <td>3D7-uganda-1</td>\n",
       "      <td>REV-1</td>\n",
       "      <td>Prism_recapture_plate3</td>\n",
       "      <td>1</td>\n",
       "      <td>I</td>\n",
       "      <td>NaN</td>\n",
       "      <td>299</td>\n",
       "      <td>deborah</td>\n",
       "      <td>DR2</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1</td>\n",
       "      <td>3D7</td>\n",
       "      <td>CTL</td>\n",
       "      <td>uganda</td>\n",
       "      <td>1</td>\n",
       "      <td>3D7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Sample ID  Library Prep 384 Column FW_plate  Original SID REV_plate  \\\n",
       "0     3D7-ROS-1        190517        Odd    FW-24     3D7-ROS-1     REV-2   \n",
       "1     3D7-ROS-2        190517        Odd    FW-44     3D7-ROS-2     REV-4   \n",
       "2     3D7-ROS-3        190517        Odd    FW-24     3D7-ROS-3     REV-2   \n",
       "3     3D7-ROS-4        190517        Odd    FW-44     3D7-ROS-4     REV-4   \n",
       "4  3D7-uganda-1        200316        Odd    FW-47  3D7-uganda-1     REV-1   \n",
       "\n",
       "            capture_plate  capture_plate_column capture_plate_row  diff   fw  \\\n",
       "0  DR2_ROS_RECAP_Plate_02                     1                 E   NaN  108   \n",
       "1  DR2_ROS_RECAP_Plate_01                     1                 M   NaN  296   \n",
       "2  DR2_ROS_RECAP_Plate_02                     3                 E   NaN   97   \n",
       "3  DR2_ROS_RECAP_Plate_01                     3                 M   NaN  297   \n",
       "4  Prism_recapture_plate3                     1                 I   NaN  299   \n",
       "\n",
       "     owner probe_set  quadrant  rev sample_name sample_plate sample_set  \\\n",
       "0  deborah       DR2       2.0   97         3D7          CTL        ROS   \n",
       "1  deborah       DR2       4.0  289         3D7          CTL        ROS   \n",
       "2  deborah       DR2       2.0   98         3D7          CTL        ROS   \n",
       "3  deborah       DR2       4.0  290         3D7          CTL        ROS   \n",
       "4  deborah       DR2       3.0    1         3D7          CTL     uganda   \n",
       "\n",
       "   replicate Sample Name  \n",
       "0          1         3D7  \n",
       "1          2         3D7  \n",
       "2          3         3D7  \n",
       "3          4         3D7  \n",
       "4          1         3D7  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "run_meta_file = os.path.join(data_dir, \"run_meta.csv\")\n",
    "run_meta = pd.read_csv(, index_col=0)\n",
    "run_meta.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Original SID</th>\n",
       "      <th>Sample ID</th>\n",
       "      <th>Library Prep</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3D7-ROS-1</td>\n",
       "      <td>3D7-ROS-1</td>\n",
       "      <td>190517</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3D7-ROS-2</td>\n",
       "      <td>3D7-ROS-2</td>\n",
       "      <td>190517</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3D7-ROS-3</td>\n",
       "      <td>3D7-ROS-3</td>\n",
       "      <td>190517</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3D7-ROS-4</td>\n",
       "      <td>3D7-ROS-4</td>\n",
       "      <td>190517</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3D7-uganda-1</td>\n",
       "      <td>3D7-uganda-1</td>\n",
       "      <td>200316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1803</th>\n",
       "      <td>TO-04-46-uganda-1</td>\n",
       "      <td>TO-04-46-uganda-1</td>\n",
       "      <td>200316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1804</th>\n",
       "      <td>TO-04-47-uganda-1</td>\n",
       "      <td>TO-04-47-uganda-1</td>\n",
       "      <td>200316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1805</th>\n",
       "      <td>TO-04-48-uganda-1</td>\n",
       "      <td>TO-04-48-uganda-1</td>\n",
       "      <td>200316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1806</th>\n",
       "      <td>TO-04-49-uganda-1</td>\n",
       "      <td>TO-04-49-uganda-1</td>\n",
       "      <td>200316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1807</th>\n",
       "      <td>TO-04-50-uganda-1</td>\n",
       "      <td>TO-04-50-uganda-1</td>\n",
       "      <td>200316</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1808 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Original SID          Sample ID  Library Prep\n",
       "0             3D7-ROS-1          3D7-ROS-1        190517\n",
       "1             3D7-ROS-2          3D7-ROS-2        190517\n",
       "2             3D7-ROS-3          3D7-ROS-3        190517\n",
       "3             3D7-ROS-4          3D7-ROS-4        190517\n",
       "4          3D7-uganda-1       3D7-uganda-1        200316\n",
       "...                 ...                ...           ...\n",
       "1803  TO-04-46-uganda-1  TO-04-46-uganda-1        200316\n",
       "1804  TO-04-47-uganda-1  TO-04-47-uganda-1        200316\n",
       "1805  TO-04-48-uganda-1  TO-04-48-uganda-1        200316\n",
       "1806  TO-04-49-uganda-1  TO-04-49-uganda-1        200316\n",
       "1807  TO-04-50-uganda-1  TO-04-50-uganda-1        200316\n",
       "\n",
       "[1808 rows x 3 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "run_meta[[\"Original SID\", \"Sample ID\", \"Library Prep\"]]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\"Dry\" merge to see if anything funny is happening such as losing some entries from the fastq table."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(7786, 10)"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fastq_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(7786, 11)"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fastq_df.merge(run_meta[[\"Original SID\", \"Sample ID\", \"Library Prep\"]]).shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Merge to add the final Sample ID to fastq table."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "fastq_df = fastq_df.merge(run_meta[[\"Original SID\", \"Sample ID\", \"Library Prep\"]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Original SID</th>\n",
       "      <th>Read Order</th>\n",
       "      <th>File Name</th>\n",
       "      <th>File Path</th>\n",
       "      <th>Run Name</th>\n",
       "      <th>Run ID</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>sample_set</th>\n",
       "      <th>replicate</th>\n",
       "      <th>Library Prep</th>\n",
       "      <th>Sample ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RXS480-ROS-1</td>\n",
       "      <td>2</td>\n",
       "      <td>RXS480-ROS-1_S953_R2_001.fastq.gz</td>\n",
       "      <td>/opt/work/190312_nextseq/fastq/RXS480-ROS-1_S9...</td>\n",
       "      <td>190312_nextseq</td>\n",
       "      <td>190312</td>\n",
       "      <td>RXS480</td>\n",
       "      <td>ROS</td>\n",
       "      <td>1</td>\n",
       "      <td>190312</td>\n",
       "      <td>RXS480-ROS-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>RXS480-ROS-1</td>\n",
       "      <td>1</td>\n",
       "      <td>RXS480-ROS-1_S953_R1_001.fastq.gz</td>\n",
       "      <td>/opt/work/190312_nextseq/fastq/RXS480-ROS-1_S9...</td>\n",
       "      <td>190312_nextseq</td>\n",
       "      <td>190312</td>\n",
       "      <td>RXS480</td>\n",
       "      <td>ROS</td>\n",
       "      <td>1</td>\n",
       "      <td>190312</td>\n",
       "      <td>RXS480-ROS-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>RXS480-ROS-1</td>\n",
       "      <td>2</td>\n",
       "      <td>RXS480-ROS-1_S1004_R2_001.fastq.gz</td>\n",
       "      <td>/opt/work/190321_nextseq/fastq/RXS480-ROS-1_S1...</td>\n",
       "      <td>190321_nextseq</td>\n",
       "      <td>190321</td>\n",
       "      <td>RXS480</td>\n",
       "      <td>ROS</td>\n",
       "      <td>1</td>\n",
       "      <td>190312</td>\n",
       "      <td>RXS480-ROS-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>RXS480-ROS-1</td>\n",
       "      <td>1</td>\n",
       "      <td>RXS480-ROS-1_S1004_R1_001.fastq.gz</td>\n",
       "      <td>/opt/work/190321_nextseq/fastq/RXS480-ROS-1_S1...</td>\n",
       "      <td>190321_nextseq</td>\n",
       "      <td>190321</td>\n",
       "      <td>RXS480</td>\n",
       "      <td>ROS</td>\n",
       "      <td>1</td>\n",
       "      <td>190312</td>\n",
       "      <td>RXS480-ROS-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>RXS480-ROS-1</td>\n",
       "      <td>2</td>\n",
       "      <td>RXS480-ROS-1_S2699_R2_001.fastq.gz</td>\n",
       "      <td>/opt/work/190405_nextseq/fastq/RXS480-ROS-1_S2...</td>\n",
       "      <td>190405_nextseq</td>\n",
       "      <td>190405</td>\n",
       "      <td>RXS480</td>\n",
       "      <td>ROS</td>\n",
       "      <td>1</td>\n",
       "      <td>190312</td>\n",
       "      <td>RXS480-ROS-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Original SID Read Order                           File Name  \\\n",
       "0  RXS480-ROS-1          2   RXS480-ROS-1_S953_R2_001.fastq.gz   \n",
       "1  RXS480-ROS-1          1   RXS480-ROS-1_S953_R1_001.fastq.gz   \n",
       "2  RXS480-ROS-1          2  RXS480-ROS-1_S1004_R2_001.fastq.gz   \n",
       "3  RXS480-ROS-1          1  RXS480-ROS-1_S1004_R1_001.fastq.gz   \n",
       "4  RXS480-ROS-1          2  RXS480-ROS-1_S2699_R2_001.fastq.gz   \n",
       "\n",
       "                                           File Path        Run Name  Run ID  \\\n",
       "0  /opt/work/190312_nextseq/fastq/RXS480-ROS-1_S9...  190312_nextseq  190312   \n",
       "1  /opt/work/190312_nextseq/fastq/RXS480-ROS-1_S9...  190312_nextseq  190312   \n",
       "2  /opt/work/190321_nextseq/fastq/RXS480-ROS-1_S1...  190321_nextseq  190321   \n",
       "3  /opt/work/190321_nextseq/fastq/RXS480-ROS-1_S1...  190321_nextseq  190321   \n",
       "4  /opt/work/190405_nextseq/fastq/RXS480-ROS-1_S2...  190405_nextseq  190405   \n",
       "\n",
       "  sample_name sample_set  replicate  Library Prep     Sample ID  \n",
       "0      RXS480        ROS          1        190312  RXS480-ROS-1  \n",
       "1      RXS480        ROS          1        190312  RXS480-ROS-1  \n",
       "2      RXS480        ROS          1        190312  RXS480-ROS-1  \n",
       "3      RXS480        ROS          1        190312  RXS480-ROS-1  \n",
       "4      RXS480        ROS          1        190312  RXS480-ROS-1  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fastq_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Rename files\n",
    "Non-unique fastq file names may be generated when multiple sequencing runs are performed for the same libraries. However, all Sample IDs for each run are unique. So we'll rename all files to make them all unique."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "fastq_df[\"New File Name\"] = (fastq_df[\"Sample ID\"] + \"-\" + fastq_df[\"Run ID\"].astype(str) \n",
    "                                 + \"_R\" + fastq_df[\"Read Order\"].astype(str) + \".fastq.gz\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Copy fastq files\n",
    "Create a fastq directory within the project specific data directory and copy all fastq files into it.  \n",
    "Inspect the process output to make sure the command completes with returncode 0 and no errors."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "CompletedProcess(args=['mkdir', '-p', '/opt/data/project_data/fastq'], returncode=0, stdout=b'', stderr=b'')"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "project_fastq_dir = os.path.join(data_dir, \"fastq\")\n",
    "subprocess.run([\"mkdir\", \"-p\", project_fastq_dir], \n",
    "               stdout=subprocess.PIPE, stderr=subprocess.PIPE)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Add new file path which is the project fastq directory and the new file name."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "fastq_df[\"New File Path\"] = fastq_df[\"New File Name\"].apply(\n",
    "    lambda a: os.path.join(project_fastq_dir, a))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Copy files using rsync."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "res = fastq_df.apply(lambda a: subprocess.run(\n",
    "    [\"rsync\", \"-a\", a[\"File Path\"], a[\"New File Path\"]],\n",
    "    stdout=subprocess.PIPE, stderr=subprocess.PIPE), axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Check that all processes completed with 0 returncode. If the below sum is greater than 0, the called process objects in the res must be inspected."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res.apply(lambda a: a.returncode).sum()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Link BioSample accession numbers\n",
    "All samples used in the study must have been submitted to the NIH BioSample data base. Place the sample metadata file in the project's data directory and load it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>organism</th>\n",
       "      <th>isolate</th>\n",
       "      <th>host</th>\n",
       "      <th>isolation_source</th>\n",
       "      <th>collection_date</th>\n",
       "      <th>geo_loc_name</th>\n",
       "      <th>sample_type</th>\n",
       "      <th>accession</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RXS1</td>\n",
       "      <td>Plasmodium falciparum</td>\n",
       "      <td>RXS1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Dried Blood Spot</td>\n",
       "      <td>2018</td>\n",
       "      <td>Uganda: Agago</td>\n",
       "      <td>Dried Blood Spot</td>\n",
       "      <td>SAMN15749034</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>RXS2</td>\n",
       "      <td>Plasmodium falciparum</td>\n",
       "      <td>RXS2</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Dried Blood Spot</td>\n",
       "      <td>2018</td>\n",
       "      <td>Uganda: Agago</td>\n",
       "      <td>Dried Blood Spot</td>\n",
       "      <td>SAMN15749035</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>RXS3</td>\n",
       "      <td>Plasmodium falciparum</td>\n",
       "      <td>RXS3</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Dried Blood Spot</td>\n",
       "      <td>2018</td>\n",
       "      <td>Uganda: Agago</td>\n",
       "      <td>Dried Blood Spot</td>\n",
       "      <td>SAMN15749036</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>RXS4</td>\n",
       "      <td>Plasmodium falciparum</td>\n",
       "      <td>RXS4</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Dried Blood Spot</td>\n",
       "      <td>2018</td>\n",
       "      <td>Uganda: Agago</td>\n",
       "      <td>Dried Blood Spot</td>\n",
       "      <td>SAMN15749037</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>RXS5</td>\n",
       "      <td>Plasmodium falciparum</td>\n",
       "      <td>RXS5</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Dried Blood Spot</td>\n",
       "      <td>2018</td>\n",
       "      <td>Uganda: Agago</td>\n",
       "      <td>Dried Blood Spot</td>\n",
       "      <td>SAMN15749038</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_name               organism isolate          host  isolation_source  \\\n",
       "0        RXS1  Plasmodium falciparum    RXS1  Homo sapiens  Dried Blood Spot   \n",
       "1        RXS2  Plasmodium falciparum    RXS2  Homo sapiens  Dried Blood Spot   \n",
       "2        RXS3  Plasmodium falciparum    RXS3  Homo sapiens  Dried Blood Spot   \n",
       "3        RXS4  Plasmodium falciparum    RXS4  Homo sapiens  Dried Blood Spot   \n",
       "4        RXS5  Plasmodium falciparum    RXS5  Homo sapiens  Dried Blood Spot   \n",
       "\n",
       "   collection_date   geo_loc_name       sample_type     accession  \n",
       "0             2018  Uganda: Agago  Dried Blood Spot  SAMN15749034  \n",
       "1             2018  Uganda: Agago  Dried Blood Spot  SAMN15749035  \n",
       "2             2018  Uganda: Agago  Dried Blood Spot  SAMN15749036  \n",
       "3             2018  Uganda: Agago  Dried Blood Spot  SAMN15749037  \n",
       "4             2018  Uganda: Agago  Dried Blood Spot  SAMN15749038  "
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_meta_sra = pd.read_table(os.path.join(data_dir, \"sra_sample_meta.tsv\"))\n",
    "sample_meta_sra.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Include the control sample metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Control Type</th>\n",
       "      <th>DNA Concentration (ng/ul)</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>organism</th>\n",
       "      <th>isolate</th>\n",
       "      <th>collected_by</th>\n",
       "      <th>collection_date</th>\n",
       "      <th>geo_loc_name</th>\n",
       "      <th>host</th>\n",
       "      <th>host_disease</th>\n",
       "      <th>isolation_source</th>\n",
       "      <th>lat_lon</th>\n",
       "      <th>accession</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>POS</td>\n",
       "      <td>1.25E-05</td>\n",
       "      <td>D0</td>\n",
       "      <td>Plasmodium Falciparum</td>\n",
       "      <td>D0</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Malaria</td>\n",
       "      <td>Laboratory</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>SAMN15747964</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>POS</td>\n",
       "      <td>2.50E-05</td>\n",
       "      <td>D1</td>\n",
       "      <td>Plasmodium Falciparum</td>\n",
       "      <td>D1</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Malaria</td>\n",
       "      <td>Laboratory</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>SAMN15747965</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>POS</td>\n",
       "      <td>5.00E-05</td>\n",
       "      <td>D2</td>\n",
       "      <td>Plasmodium Falciparum</td>\n",
       "      <td>D2</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Malaria</td>\n",
       "      <td>Laboratory</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>SAMN15747966</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>POS</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>D3</td>\n",
       "      <td>Plasmodium Falciparum</td>\n",
       "      <td>D3</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Malaria</td>\n",
       "      <td>Laboratory</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>SAMN15747967</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>POS</td>\n",
       "      <td>0.000175</td>\n",
       "      <td>D4</td>\n",
       "      <td>Plasmodium Falciparum</td>\n",
       "      <td>D4</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Malaria</td>\n",
       "      <td>Laboratory</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>SAMN15747968</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Control Type DNA Concentration (ng/ul) sample_name               organism  \\\n",
       "0          POS                  1.25E-05          D0  Plasmodium Falciparum   \n",
       "1          POS                  2.50E-05          D1  Plasmodium Falciparum   \n",
       "2          POS                  5.00E-05          D2  Plasmodium Falciparum   \n",
       "3          POS                    0.0001          D3  Plasmodium Falciparum   \n",
       "4          POS                  0.000175          D4  Plasmodium Falciparum   \n",
       "\n",
       "  isolate    collected_by collection_date    geo_loc_name          host  \\\n",
       "0      D0  not applicable  not applicable  not applicable  Homo sapiens   \n",
       "1      D1  not applicable  not applicable  not applicable  Homo sapiens   \n",
       "2      D2  not applicable  not applicable  not applicable  Homo sapiens   \n",
       "3      D3  not applicable  not applicable  not applicable  Homo sapiens   \n",
       "4      D4  not applicable  not applicable  not applicable  Homo sapiens   \n",
       "\n",
       "  host_disease isolation_source         lat_lon     accession  \n",
       "0      Malaria       Laboratory  not applicable  SAMN15747964  \n",
       "1      Malaria       Laboratory  not applicable  SAMN15747965  \n",
       "2      Malaria       Laboratory  not applicable  SAMN15747966  \n",
       "3      Malaria       Laboratory  not applicable  SAMN15747967  \n",
       "4      Malaria       Laboratory  not applicable  SAMN15747968  "
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "control_meta_sra = pd.read_table(os.path.join(data_dir, \"parasite_control_meta_sra.tsv\"))\n",
    "control_meta_sra.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1443, 2)"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_meta = pd.concat([sample_meta_sra, control_meta_sra], ignore_index=True)\n",
    "sample_meta = sample_meta[[\"sample_name\" ,\"accession\"]]\n",
    "sample_meta.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_meta.rename(columns={\"accession\": \"biosample_accession\"},\n",
    "                   inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Dry merge"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(7786, 13)"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fastq_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(7786, 14)"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fastq_df.merge(sample_meta).shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Merge"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "fastq_df = fastq_df.merge(sample_meta)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Convert the fastq table to SRA format\n",
    "SRA submission requires each library on a single row and all fastq files for a specific library to be listed as filename, filename2, filename3, etc.  \n",
    "\n",
    "We'll start with getting the maximum number of files per library."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "6"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "max_file_num = fastq_df.groupby(\"Sample ID\").size().max()\n",
    "max_file_num"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Define a short function that takes a groupby object and creates a single table row containing all files for the library and empty strings for libraries that don't have as many as max number of files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_files(g):\n",
    "    files = g[\"New File Name\"].tolist()\n",
    "    count = len(files)\n",
    "    empty = max_file_num - count\n",
    "    cols = [\"filename\"] + [\"filename\" + str(i) for i in range(2, max_file_num+1)]\n",
    "    for i in range(empty):\n",
    "        files.append(\"\")\n",
    "    return pd.Series(files, index=cols)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create a files table in SRA format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1787, 9)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample ID</th>\n",
       "      <th>biosample_accession</th>\n",
       "      <th>Library Prep</th>\n",
       "      <th>filename</th>\n",
       "      <th>filename2</th>\n",
       "      <th>filename3</th>\n",
       "      <th>filename4</th>\n",
       "      <th>filename5</th>\n",
       "      <th>filename6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3D7-ROS-1</td>\n",
       "      <td>SAMN15747989</td>\n",
       "      <td>190517</td>\n",
       "      <td>3D7-ROS-1-190517_R1.fastq.gz</td>\n",
       "      <td>3D7-ROS-1-190517_R2.fastq.gz</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3D7-ROS-2</td>\n",
       "      <td>SAMN15747989</td>\n",
       "      <td>190517</td>\n",
       "      <td>3D7-ROS-2-190517_R2.fastq.gz</td>\n",
       "      <td>3D7-ROS-2-190517_R1.fastq.gz</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3D7-ROS-3</td>\n",
       "      <td>SAMN15747989</td>\n",
       "      <td>190517</td>\n",
       "      <td>3D7-ROS-3-190517_R2.fastq.gz</td>\n",
       "      <td>3D7-ROS-3-190517_R1.fastq.gz</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3D7-ROS-4</td>\n",
       "      <td>SAMN15747989</td>\n",
       "      <td>190517</td>\n",
       "      <td>3D7-ROS-4-190517_R2.fastq.gz</td>\n",
       "      <td>3D7-ROS-4-190517_R1.fastq.gz</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3D7-uganda-1</td>\n",
       "      <td>SAMN15747989</td>\n",
       "      <td>200316</td>\n",
       "      <td>3D7-uganda-1-200316_R2.fastq.gz</td>\n",
       "      <td>3D7-uganda-1-200316_R1.fastq.gz</td>\n",
       "      <td>3D7-uganda-1-200609_R1.fastq.gz</td>\n",
       "      <td>3D7-uganda-1-200609_R2.fastq.gz</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Sample ID biosample_accession  Library Prep  \\\n",
       "0     3D7-ROS-1        SAMN15747989        190517   \n",
       "1     3D7-ROS-2        SAMN15747989        190517   \n",
       "2     3D7-ROS-3        SAMN15747989        190517   \n",
       "3     3D7-ROS-4        SAMN15747989        190517   \n",
       "4  3D7-uganda-1        SAMN15747989        200316   \n",
       "\n",
       "                          filename                        filename2  \\\n",
       "0     3D7-ROS-1-190517_R1.fastq.gz     3D7-ROS-1-190517_R2.fastq.gz   \n",
       "1     3D7-ROS-2-190517_R2.fastq.gz     3D7-ROS-2-190517_R1.fastq.gz   \n",
       "2     3D7-ROS-3-190517_R2.fastq.gz     3D7-ROS-3-190517_R1.fastq.gz   \n",
       "3     3D7-ROS-4-190517_R2.fastq.gz     3D7-ROS-4-190517_R1.fastq.gz   \n",
       "4  3D7-uganda-1-200316_R2.fastq.gz  3D7-uganda-1-200316_R1.fastq.gz   \n",
       "\n",
       "                         filename3                        filename4 filename5  \\\n",
       "0                                                                               \n",
       "1                                                                               \n",
       "2                                                                               \n",
       "3                                                                               \n",
       "4  3D7-uganda-1-200609_R1.fastq.gz  3D7-uganda-1-200609_R2.fastq.gz             \n",
       "\n",
       "  filename6  \n",
       "0            \n",
       "1            \n",
       "2            \n",
       "3            \n",
       "4            "
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "files = fastq_df.groupby([\"Sample ID\", \"biosample_accession\", \"Library Prep\"]).apply(get_files).reset_index()\n",
    "print(files.shape)\n",
    "files.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### SRA metadata\n",
    "SRA requires certain metadata. Most can be left as is below, but a description of the project, title, and most importantly previously generated **BioProject ID** is required."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bioproject_accession</th>\n",
       "      <th>title</th>\n",
       "      <th>library_strategy</th>\n",
       "      <th>library_source</th>\n",
       "      <th>library_selection</th>\n",
       "      <th>library_layout</th>\n",
       "      <th>platform</th>\n",
       "      <th>instrument_model</th>\n",
       "      <th>design_description</th>\n",
       "      <th>filetype</th>\n",
       "      <th>assembly</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PRJNA655702</td>\n",
       "      <td>Plasmodium falciparum targeted sequencing of d...</td>\n",
       "      <td>Targeted-Capture</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>Padlock probes capture method</td>\n",
       "      <td>Paired</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>NextSeq 550</td>\n",
       "      <td>Plasmodium falciparum drug resistance loci in ...</td>\n",
       "      <td>fastq</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  bioproject_accession                                              title  \\\n",
       "0          PRJNA655702  Plasmodium falciparum targeted sequencing of d...   \n",
       "\n",
       "   library_strategy library_source              library_selection  \\\n",
       "0  Targeted-Capture        GENOMIC  Padlock probes capture method   \n",
       "\n",
       "  library_layout  platform instrument_model  \\\n",
       "0         Paired  ILLUMINA      NextSeq 550   \n",
       "\n",
       "                                  design_description filetype assembly  \n",
       "0  Plasmodium falciparum drug resistance loci in ...    fastq           "
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "desc_string = (\"Plasmodium falciparum drug resistance loci in samples \"\n",
    "               \"from Uganda collected in 2018-2019 were \"\n",
    "               \"captured using molecular inversion probes and sequenced using \"\n",
    "               \"Illumina NextSeq platform.\")\n",
    "title = \"Plasmodium falciparum targeted sequencing of drug resistance loci in Uganda.\"\n",
    "bioproject_accession = \"PRJNA655702\"\n",
    "instrument_model = \"NextSeq 550\"\n",
    "sra_meta = {\n",
    "    \"bioproject_accession\": bioproject_accession,\n",
    "    \"title\": title,\n",
    "    \"library_strategy\": \"Targeted-Capture\",\n",
    "    \"library_source\": \"GENOMIC\",\n",
    "    \"library_selection\": \"Padlock probes capture method\",\n",
    "    \"library_layout\": \"Paired\",\n",
    "    \"platform\": \"ILLUMINA\",\n",
    "    \"instrument_model\": instrument_model,\n",
    "    \"design_description\": desc_string,\n",
    "    \"filetype\": \"fastq\",\n",
    "    \"assembly\": \"\"\n",
    "    }\n",
    "sra_meta_df = pd.DataFrame(sra_meta, index = [0])\n",
    "sra_meta_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Add the sra metadata to the files table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1787, 20)"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sra_meta_df[\"Temp\"] = \"Temp\"\n",
    "files[\"Temp\"] = \"Temp\"\n",
    "sra_meta_df.merge(files).drop(\"Temp\", axis=1).shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bioproject_accession</th>\n",
       "      <th>title</th>\n",
       "      <th>library_strategy</th>\n",
       "      <th>library_source</th>\n",
       "      <th>library_selection</th>\n",
       "      <th>library_layout</th>\n",
       "      <th>platform</th>\n",
       "      <th>instrument_model</th>\n",
       "      <th>design_description</th>\n",
       "      <th>filetype</th>\n",
       "      <th>assembly</th>\n",
       "      <th>Sample ID</th>\n",
       "      <th>biosample_accession</th>\n",
       "      <th>Library Prep</th>\n",
       "      <th>filename</th>\n",
       "      <th>filename2</th>\n",
       "      <th>filename3</th>\n",
       "      <th>filename4</th>\n",
       "      <th>filename5</th>\n",
       "      <th>filename6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PRJNA655702</td>\n",
       "      <td>Plasmodium falciparum targeted sequencing of d...</td>\n",
       "      <td>Targeted-Capture</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>Padlock probes capture method</td>\n",
       "      <td>Paired</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>NextSeq 550</td>\n",
       "      <td>Plasmodium falciparum drug resistance loci in ...</td>\n",
       "      <td>fastq</td>\n",
       "      <td></td>\n",
       "      <td>3D7-ROS-1</td>\n",
       "      <td>SAMN15747989</td>\n",
       "      <td>190517</td>\n",
       "      <td>3D7-ROS-1-190517_R1.fastq.gz</td>\n",
       "      <td>3D7-ROS-1-190517_R2.fastq.gz</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PRJNA655702</td>\n",
       "      <td>Plasmodium falciparum targeted sequencing of d...</td>\n",
       "      <td>Targeted-Capture</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>Padlock probes capture method</td>\n",
       "      <td>Paired</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>NextSeq 550</td>\n",
       "      <td>Plasmodium falciparum drug resistance loci in ...</td>\n",
       "      <td>fastq</td>\n",
       "      <td></td>\n",
       "      <td>3D7-ROS-2</td>\n",
       "      <td>SAMN15747989</td>\n",
       "      <td>190517</td>\n",
       "      <td>3D7-ROS-2-190517_R2.fastq.gz</td>\n",
       "      <td>3D7-ROS-2-190517_R1.fastq.gz</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PRJNA655702</td>\n",
       "      <td>Plasmodium falciparum targeted sequencing of d...</td>\n",
       "      <td>Targeted-Capture</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>Padlock probes capture method</td>\n",
       "      <td>Paired</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>NextSeq 550</td>\n",
       "      <td>Plasmodium falciparum drug resistance loci in ...</td>\n",
       "      <td>fastq</td>\n",
       "      <td></td>\n",
       "      <td>3D7-ROS-3</td>\n",
       "      <td>SAMN15747989</td>\n",
       "      <td>190517</td>\n",
       "      <td>3D7-ROS-3-190517_R2.fastq.gz</td>\n",
       "      <td>3D7-ROS-3-190517_R1.fastq.gz</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PRJNA655702</td>\n",
       "      <td>Plasmodium falciparum targeted sequencing of d...</td>\n",
       "      <td>Targeted-Capture</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>Padlock probes capture method</td>\n",
       "      <td>Paired</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>NextSeq 550</td>\n",
       "      <td>Plasmodium falciparum drug resistance loci in ...</td>\n",
       "      <td>fastq</td>\n",
       "      <td></td>\n",
       "      <td>3D7-ROS-4</td>\n",
       "      <td>SAMN15747989</td>\n",
       "      <td>190517</td>\n",
       "      <td>3D7-ROS-4-190517_R2.fastq.gz</td>\n",
       "      <td>3D7-ROS-4-190517_R1.fastq.gz</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PRJNA655702</td>\n",
       "      <td>Plasmodium falciparum targeted sequencing of d...</td>\n",
       "      <td>Targeted-Capture</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>Padlock probes capture method</td>\n",
       "      <td>Paired</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>NextSeq 550</td>\n",
       "      <td>Plasmodium falciparum drug resistance loci in ...</td>\n",
       "      <td>fastq</td>\n",
       "      <td></td>\n",
       "      <td>3D7-uganda-1</td>\n",
       "      <td>SAMN15747989</td>\n",
       "      <td>200316</td>\n",
       "      <td>3D7-uganda-1-200316_R2.fastq.gz</td>\n",
       "      <td>3D7-uganda-1-200316_R1.fastq.gz</td>\n",
       "      <td>3D7-uganda-1-200609_R1.fastq.gz</td>\n",
       "      <td>3D7-uganda-1-200609_R2.fastq.gz</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  bioproject_accession                                              title  \\\n",
       "0          PRJNA655702  Plasmodium falciparum targeted sequencing of d...   \n",
       "1          PRJNA655702  Plasmodium falciparum targeted sequencing of d...   \n",
       "2          PRJNA655702  Plasmodium falciparum targeted sequencing of d...   \n",
       "3          PRJNA655702  Plasmodium falciparum targeted sequencing of d...   \n",
       "4          PRJNA655702  Plasmodium falciparum targeted sequencing of d...   \n",
       "\n",
       "   library_strategy library_source              library_selection  \\\n",
       "0  Targeted-Capture        GENOMIC  Padlock probes capture method   \n",
       "1  Targeted-Capture        GENOMIC  Padlock probes capture method   \n",
       "2  Targeted-Capture        GENOMIC  Padlock probes capture method   \n",
       "3  Targeted-Capture        GENOMIC  Padlock probes capture method   \n",
       "4  Targeted-Capture        GENOMIC  Padlock probes capture method   \n",
       "\n",
       "  library_layout  platform instrument_model  \\\n",
       "0         Paired  ILLUMINA      NextSeq 550   \n",
       "1         Paired  ILLUMINA      NextSeq 550   \n",
       "2         Paired  ILLUMINA      NextSeq 550   \n",
       "3         Paired  ILLUMINA      NextSeq 550   \n",
       "4         Paired  ILLUMINA      NextSeq 550   \n",
       "\n",
       "                                  design_description filetype assembly  \\\n",
       "0  Plasmodium falciparum drug resistance loci in ...    fastq            \n",
       "1  Plasmodium falciparum drug resistance loci in ...    fastq            \n",
       "2  Plasmodium falciparum drug resistance loci in ...    fastq            \n",
       "3  Plasmodium falciparum drug resistance loci in ...    fastq            \n",
       "4  Plasmodium falciparum drug resistance loci in ...    fastq            \n",
       "\n",
       "      Sample ID biosample_accession  Library Prep  \\\n",
       "0     3D7-ROS-1        SAMN15747989        190517   \n",
       "1     3D7-ROS-2        SAMN15747989        190517   \n",
       "2     3D7-ROS-3        SAMN15747989        190517   \n",
       "3     3D7-ROS-4        SAMN15747989        190517   \n",
       "4  3D7-uganda-1        SAMN15747989        200316   \n",
       "\n",
       "                          filename                        filename2  \\\n",
       "0     3D7-ROS-1-190517_R1.fastq.gz     3D7-ROS-1-190517_R2.fastq.gz   \n",
       "1     3D7-ROS-2-190517_R2.fastq.gz     3D7-ROS-2-190517_R1.fastq.gz   \n",
       "2     3D7-ROS-3-190517_R2.fastq.gz     3D7-ROS-3-190517_R1.fastq.gz   \n",
       "3     3D7-ROS-4-190517_R2.fastq.gz     3D7-ROS-4-190517_R1.fastq.gz   \n",
       "4  3D7-uganda-1-200316_R2.fastq.gz  3D7-uganda-1-200316_R1.fastq.gz   \n",
       "\n",
       "                         filename3                        filename4 filename5  \\\n",
       "0                                                                               \n",
       "1                                                                               \n",
       "2                                                                               \n",
       "3                                                                               \n",
       "4  3D7-uganda-1-200609_R1.fastq.gz  3D7-uganda-1-200609_R2.fastq.gz             \n",
       "\n",
       "  filename6  \n",
       "0            \n",
       "1            \n",
       "2            \n",
       "3            \n",
       "4            "
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sra_meta_df = sra_meta_df.merge(files).drop(\"Temp\", axis=1)\n",
    "sra_meta_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Rename Sample ID to library_ID as per SRA format and remove the now unnecessary Library Prep column."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [],
   "source": [
    "sra_meta_df.rename(\n",
    "    columns={\"Sample ID\": \"library_ID\"}, inplace=True)\n",
    "sra_meta_df.drop(\"Library Prep\", axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Split the submission to multiple submissions each of which has  <1000 libraries \n",
    "Yes, the submission is limited to 999 samples. So we'll create a meta data for each \"sub\"submission and move files for that submission to a new location below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [],
   "source": [
    "mv_results = []\n",
    "mkdir_results = []\n",
    "for i in range(((sra_meta_df.shape[0] - 1) // 999) + 1):\n",
    "    start_index = i * 999\n",
    "    end_index = start_index + 999\n",
    "    sra_meta_df.iloc[start_index:end_index].to_csv(\n",
    "        os.path.join(data_dir, \"sra_meta_\" + str(i+1) + \".tsv\"), sep=\"\\t\", index=False)\n",
    "    new_fastq_dir = os.path.join(project_fastq_dir, str(i+1))\n",
    "    res = subprocess.run([\"mkdir\", new_fastq_dir],\n",
    "                         stdout=subprocess.PIPE, stderr=subprocess.PIPE)\n",
    "    mkdir_results.append(res)\n",
    "    res = fastq_df.loc[fastq_df[\"Sample ID\"].isin(sra_meta_df.iloc[start_index:end_index][\"library_ID\"]),\n",
    "            \"New File Path\"].apply(lambda a: subprocess.run(\n",
    "                                    [\"mv\", a, new_fastq_dir],\n",
    "                                    stdout=subprocess.PIPE, stderr=subprocess.PIPE))\n",
    "    mv_results.append(res)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Check the subprocess outputs to make sure there are no errors. We expect the returncode sums to be 0."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for res in mv_results:\n",
    "    print(res.apply(lambda a: a.returncode).sum())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for res in mkdir_results:\n",
    "    print(res.returncode)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we have generated an SRA metadata file in the project directory, a fastq file directory within the project fastq file directory for **each sub-submission**. The fastqs will be in \"/opt/data/project_data/fastq/1\" for the first sub-submission, for example.\n",
    "\n",
    "Let's look at the submission procedure again"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# NIH SRA submission\n",
    "This notebook was used to create an NIH SRA (fastq) submission for Plasmodium falciparum MIP sequencing data in collaboration with Rosenthal Lab. It will also serve as a reference for falciparum sample submissions. \n",
    "\n",
    "### Prerequisites\n",
    "1) Prior to SRA submission, create a bioproject on https://submit.ncbi.nlm.nih.gov/subs/bioproject/  \n",
    "This is a simple procedure that produces a bioproject ID which will be associated with one or more SRA submissions.  \n",
    "2) Prior to SRA submission, submit all of the samples used in a project to NIH BioSample database.\n",
    "\n",
    "\n",
    "\n",
    "SRA submissions are made through the NIH website: https://submit.ncbi.nlm.nih.gov/subs/sra\n",
    "\n",
    "There are 5 steps in the submission portal for SRA submissions that are listed below. **The main purpose of this notebook is to create the file needed at step 4**.\n",
    "\n",
    "Step 1: Submitter details\n",
    "\n",
    "Step 2: General Info\n",
    "  * Did you already register a BioProject for this research, e.g. for the submission of the reads to SRA: **YES**\n",
    "  * Did you already register a BioSample for this sample, e.g. for the submission of the reads to SRA: **YES**\n",
    "  * Chose data release time (immediate, or future): we will chose a future date during the sumbission. Once the sumbission is finished, an email to SRA requesting removal of possible human sequences is recommended. Although our targeted sequencing should not generate any human data, and even if there are some off targets, those would not have any identifying information. Still, SRA provides this service and it makes sense to take it. Once that is done, you can publish the SRA publicly, or wait until publication date.\n",
    "  \n",
    "  \n",
    "Step 3: SRA metadata\n",
    "  * Upload a file using Excel or text format (tab-delimited): This file will be created using this notebook.\n",
    "  \n",
    "Step 4: Files\n",
    "  * This notebook will copy the fastq files to specific directories. The files will be uploaded to the FTP server of NIH at this step.\n",
    "  * The first step in the file transfer is to use command line and navigate to wherever the fastq files were saved (e.g. ssh to seekdeep server and cd to the fastq directory). Optionally, start a screen if you are comfortable using one. I'll call this location the \"data location\".\n",
    "  \n",
    "  * On the SRA submission portal choose: *FTP or Aspera Command Line file preload*\n",
    "  * click on FTP upload instructions\n",
    "  \n",
    "  * From the *data location* connect to the NIH FTP server using the credentials provided on the submission portal\n",
    "    * ftp ftp-private.ncbi.nlm.nih.gov\n",
    "    * follow instructions on the portal until step 6\n",
    "    * for step 6, instead do the following commands\n",
    "    * prompt (this should turn off interactive, if it was already off, then it'll turn it on. We want this off, so if needed repeat the \"prompt\" command to turn off interactive.\n",
    "    * mput *\n",
    "  * Once the upload is complete go back to submission portal, click **Select preload folder**\n",
    "  * Select the folder containing the uploaded files.\n",
    "  * Autofinish\n",
    "  * Continue\n",
    "  \n",
    "\n",
    "Step 5: Review and submit."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Upload SRA metadata\n",
    "We'll pick up from step 3 where we will upload the meta files generated for each submission."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Upload fastq files\n",
    "Following the instuctions in step 4, we'll:\n",
    "  * navigate to \"/opt/data/project_data/fastq/1\"\n",
    "  * start a screen\n",
    "  * connect to NIH ftp\n",
    "  * upload the files"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
