{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "6b220bad",
   "metadata": {},
   "source": [
    "# Calling genotypes, prevalences and filtering data\n",
    "This program analyzes tables of counts. In the prior \"variant calling\" step of the program, mutation counts were generated for each sample by annotating the VCF file generated by freebayes, using SNPEff to assign amino acid names to mutations.\n",
    "\n",
    "The count tables are intended to be as \"raw\" as possible. In this part of the analysis we will:\n",
    "1. apply various filters to the count tables, considering a mutation 'sampled' for a sample if it passes the filters for that sample.\n",
    "2. choose the level of detail of geographic locations we care about from a metadata file\n",
    "3. summarize the number of filtered samples at each location that had the mutation and the total number of filtered samples at each location, where the 'prevalence' of a mutation at a location is the first number divided by the second number\n",
    "4. graph the prevalences and filtered sample sizes on a map\n",
    "5. (optional) - apply additional filtering for re-evaluation of prevalence data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "69474ffc-3f7f-48b3-b0e3-86a8eeba24fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# RUN\n",
    "import sys\n",
    "sys.path.append(\"/opt/src\")\n",
    "import calculate_prevalences as cap\n",
    "import os\n",
    "import pandas as pd\n",
    "import plotly.express as px\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "wdir = \"/opt/user/stats_and_variant_calling/\"\n",
    "import ipywidgets as widgets\n",
    "import yaml"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4a6956d-8904-428d-b6ae-01a03670c347",
   "metadata": {},
   "outputs": [],
   "source": [
    "# RUN\n",
    "\n",
    "mutation_count_file = \"alternate_AA_table.csv\"\n",
    "mutation_coverage_file = \"coverage_AA_table.csv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a977714e-c076-448e-a023-93526436773d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# RUN\n",
    "import pandas as pd\n",
    "mutation_counts = pd.read_csv(mutation_count_file,\n",
    "                              header=list(range(6)),\n",
    "                              index_col=0)\n",
    "mutation_coverage = pd.read_csv(mutation_coverage_file,\n",
    "                                index_col=0,\n",
    "                                header=list(range(6)))\n",
    "\n",
    "metadata_table=pd.read_csv('/opt/user/prevalence_metadata/metadata.csv')\n",
    "\n",
    "mutation_handle=open(mutation_count_file)\n",
    "mutation_handle.readline()\n",
    "mutation_handle.readline()\n",
    "mutations=mutation_handle.readline().strip().split(',')\n",
    "filtered_mutations=[]\n",
    "for mutation in mutations[1:]:\n",
    "    change=mutation.split('-')[-1]\n",
    "    if 'ins' not in change and 'del' not in change and 'dup' not in change:\n",
    "        filtered_mutations.append(mutation)\n",
    "print('observed mutations are:\\n\\n')\n",
    "print(filtered_mutations)\n",
    "print('\\n\\nmetadata table is')\n",
    "metadata_table.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b05fb266",
   "metadata": {},
   "outputs": [],
   "source": [
    "# USER INPUT\n",
    "\n",
    "# how many reads (or, more accurately, UMIs) do we need spanning the genomic coordinates of a mutation in order to\n",
    "# consider the coordinates \"sequenced\" for a given sample? If UMI count is below min_coverage, the sample will be\n",
    "# marked as \"missing\" for the mutation of interest.\n",
    "min_coverage = 10\n",
    "\n",
    "# If min_coverage is achieved in the sample, how many UMIs do we need that support the mutant (or alternate) allele\n",
    "# in order to consider the alternate allele \"present\" in the sample? Otherwise if min_coverage is surpassed but\n",
    "# min_count is not, the sample will be marked as matching the reference allele.\n",
    "min_count = 3\n",
    "\n",
    "# Is there a number that min_count/min_coverage needs to surpass before we consider an alternate allele \"present\" in a\n",
    "# sample? Our usual answer is 'no' so we generally set this to '0' - but you may not want to count an alternate allele\n",
    "# present if it's found in a very low fraction of the UMIs associated with a sample, even if the absolute number of UMIs\n",
    "# associated with the alternate allele is above your threshold.\n",
    "min_freq = 0\n",
    "\n",
    "# Which mutations would you like to graph? A full list of sequenced plus key mutations is created when you run the code in the\n",
    "# cell above.\n",
    "\n",
    "#example if you want only \"key mutations from the targets.tsv file found in the project_resources folder\"\n",
    "mutation_column=open('/opt/project_resources/targets.tsv').readline().strip().split('\\t').index('mutation_name')\n",
    "mutations_of_interest=[line.strip().split('\\t')[mutation_column] for line in open('/opt/project_resources/targets.tsv')][1:]\n",
    "\n",
    "#example if you want only mutations from the AA tables from a specific gene\n",
    "#mutations_of_interest=[mutation for mutation in filtered_mutations if 'k13' in mutation]\n",
    "\n",
    "#example if you want to manually set the mutations to analyze\n",
    "#mutations_of_interest=[\n",
    "#'dhfr-ts-Asn51Ile', 'dhfr-ts-Cys59Arg', 'dhfr-ts-Ser108Asn', 'dhfr-ts-Ile164Leu',\n",
    "#'mdr1-Asn652Asp', 'mdr1-Asn86Tyr', 'mdr1-Asp1246Tyr', 'mdr1-Asp650Asn',\n",
    "#'mdr1-Tyr184Phe', 'crt-Met74Ile', 'crt-Asn75Glu', 'crt-Lys76Thr',\n",
    "#'crt-Gln271Glu', 'crt-Arg371Ile', 'dhps-Ala437Gly', 'dhps-Lys540Glu',\n",
    "#'dhps-Ala581Gly']\n",
    "\n",
    "\n",
    "# In the metadata sheet (reproduced in the table output from the cell above), which column contains sample names and\n",
    "# which column contains the names of the locations you'd like to plot (or summarize by)?\n",
    "sample_column='Sampleid'\n",
    "summary_column='Sites'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f6325660-4020-43c5-80e9-8fe37bb95463",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Which country would you like to graph?\n",
    "country_shortcuts={'DRC': \n",
    "                   {\"lat\": -1.6815695315287824,\n",
    "                    \"lon\": 22.744896416745945,\n",
    "                    'zoom': 3.8},\n",
    "                   'Uganda': \n",
    "                   {\"lat\": 1.5,\n",
    "                    'lon': 32,\n",
    "                    'zoom': 5.6},\n",
    "                   'Tanzania':\n",
    "                   {\"lat\": -5.7,\n",
    "                    \"lon\": 35,\n",
    "                    'zoom': 5.2},\n",
    "                  }\n",
    "\n",
    "country = widgets.Dropdown(\n",
    "    options=country_shortcuts.keys(),\n",
    "    description='Country:',\n",
    "    disabled=False,\n",
    ")\n",
    "print(\"Country to zoom in and center the map on:\\n\")\n",
    "display(country)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3b8f2ab5-2020-44cd-85b5-e60c195bf42d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# RUN\n",
    "country=country.value\n",
    "zoom_level=country_shortcuts[country]['zoom']\n",
    "latitude=country_shortcuts[country]['lat']\n",
    "longitude=country_shortcuts[country]['lon']\n",
    "# OPTIONAL USER INPUT:\n",
    "# You can override the values above and set a custom zoom level and centering below by commenting these lines in\n",
    "#zoom_level=4.8\n",
    "#latitude=1.5\n",
    "#longitude=32"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c87661e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# RUN\n",
    "\n",
    "# import the PCA module which has genotype calling and\n",
    "# filtering functions\n",
    "import PCA\n",
    "\n",
    "gt_calls = PCA.call_genotypes(mutation_counts, mutation_coverage,\n",
    "                              min_count, min_coverage, min_freq)\n",
    "#print('options are', gt_calls.keys())\n",
    "\n",
    "# This step outputs samples that surpass the coverage threshold\n",
    "# as their original coverage value and resets samples not passing the threshold as '0'.\n",
    "# Note that even if there is insufficient coverage the resulting cell will be a '0' and not NaN\n",
    "filtered_coverage=gt_calls[\"filtered_mutation_coverage\"]\n",
    "filtered_coverage.to_csv(os.path.join(wdir, \"filtered_coverage_AA_table.csv\"))\n",
    "\n",
    "# This step outputs samples containing alternate alleles that surpass the coverage and alternate thresholds\n",
    "# as their original values and samples not passing both thresholds as '0'. Note that even if there is insufficient\n",
    "# coverage to make a call the output for this table will still be listed as '0'\n",
    "filtered_mutation_counts = gt_calls[\"filtered_mutation_counts\"]\n",
    "filtered_mutation_counts.to_csv(os.path.join(\n",
    "        wdir, \"filtered_alternate_AA_table.csv\"))\n",
    "\n",
    "# This step outputs the within sample allele frequencies of each alternate allele. Unlike the tables above, this\n",
    "# one displays NaN if the coverage levels do not surpass the threshold (due to dividing alternate counts by 0)\n",
    "freq = gt_calls[\"wsaf\"]\n",
    "freq.to_csv(os.path.join(\n",
    "        wdir, \"within_sample_allele_frequencies.csv\"))\n",
    "freq.head()\n",
    "\n",
    "# This step outputs a genotypes table, using the wsaf table as input. Cells can be NaN (coverage is below the\n",
    "# threshold), 0 (frequency of the mutation is less than or equal to min_freq), 1 (frequency is bigger than min_freq but\n",
    "# less than 100%) or 2 (mutation is found in all of the UMIs for the sample.\n",
    "genotypes = gt_calls[\"genotypes\"]\n",
    "genotypes.to_csv(os.path.join(\n",
    "        wdir, \"filtered_genotypes_table.csv\"))\n",
    "\n",
    "\n",
    "# The outputs of this table are the same as the genotypes table, except this table does not differentiate between\n",
    "# alternate alleles that are found in all UMIs of a sample and those that are found in some of the UMIs\n",
    "# (both '2' and '1' values from the genotypes table get reset to '1')\n",
    "prevalences_input = gt_calls[\"prevalences\"]\n",
    "prevalences_input.to_csv(os.path.join(wdir, \"prevalences_input_table.csv\"))\n",
    "\n",
    "\n",
    "# To view the outputs of any given step, comment in the corresponding table below (or open the corresponding csv\n",
    "# file in a spreadsheet):\n",
    "# filtered_coverage.head()\n",
    "# filtered_mutation_counts.head()\n",
    "# genotypes.head()\n",
    "# prevalences_input.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "153c029e",
   "metadata": {},
   "source": [
    "# Calculate Prevalence in Each Region"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45f4b571-2402-4766-a1cf-5baecd7879e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# RUN\n",
    "# Intersects the variants of interest with the table header\n",
    "# default sort will sort mutations by gene and then reference amino acid name - this sorter sorts by gene and\n",
    "# then amino acid position\n",
    "def special_sort(mutations):\n",
    "    sorting_list=[]\n",
    "    for mutation in mutations:\n",
    "        gene='-'.join(mutation.split('-')[:-1])\n",
    "        pos=int(mutation.split('-')[-1][3:-3])\n",
    "        sorting_list.append([gene, pos, mutation])\n",
    "    return [item[-1] for item in sorted(sorting_list)]\n",
    "\n",
    "prevalences_input='prevalences_input_table.csv'\n",
    "variant_set = set(open(prevalences_input).readlines()[2].strip().split(',')[1:])\n",
    "mutations_of_interest_set = set(mutations_of_interest)\n",
    "variants_of_interest = list(variant_set.intersection(mutations_of_interest_set))\n",
    "variants_of_interest=special_sort(mutations_of_interest)\n",
    "print(variants_of_interest)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d754547d-b88f-4315-bad8-c9078a4e1abf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# RUN\n",
    "# This block loads the config file to find the associated metadata for your sample set. It summarizes the dataset by\n",
    "# the user-specified summary_column and counts the number of '1' and '0' values from the prevalences_input_table\n",
    "# for each named location found in the summary column. e.g. if there are 5 regions in a country and we're summarizing\n",
    "# by region, there will be 5 named locations in the output table, and we will count the number of 'covered' samples in\n",
    "# each region for each mutation along with the number of samples with the alternate allele for each mutation. These\n",
    "# values appear as numerators and denominators in the output table below. Coordinates are averaged across all samples\n",
    "# that share the same named location in the metadata. All sample sets need to share a single metadata file - this is to\n",
    "# avoid naming conflicts in column names between different sample sets.\n",
    "config_file = open('/opt/config/config_v0.5.1.yaml')\n",
    "config_dict = yaml.safe_load(config_file)\n",
    "\n",
    "prevalences_input_table = 'prevalences_input_table.csv'\n",
    "\n",
    "contents=os.listdir('/opt/user/prevalence_metadata')\n",
    "print('contents are', contents)\n",
    "\n",
    "# get the sample sets from the user defined config file\n",
    "sample_sets = config_dict['sample_set'].split(',')\n",
    "if '' in sample_sets:\n",
    "    sample_sets.remove('')\n",
    "sample_sets = [x.replace(' ','') for x in sample_sets]\n",
    "\n",
    "#path to the metadata - this is currently hardcoded\n",
    "metadata_file = '/opt/user/prevalence_metadata/metadata.csv'\n",
    "\n",
    "# generate a prevalence table for each set of samples\n",
    "output_summary_table = 'prevalence_summary.tsv'\n",
    "\n",
    "#    metadata_file='merged_corrected_renamed_samples_metadata.csv'\n",
    "cap.calculate_prevalences(metadata_file,\n",
    "                          prevalences_input_table,\n",
    "                          mutations_of_interest,\n",
    "                          output_summary_table, sample_column, summary_column)\n",
    "prevalences=pd.read_csv(output_summary_table,\n",
    "                          header=list(range(1)),\n",
    "                          index_col=0, sep='\\t')\n",
    "prevalences.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "83a1a388",
   "metadata": {},
   "source": [
    "# Plot Scatter map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7492699f-f927-49da-a039-b33b8914eee7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# RUN\n",
    "import math\n",
    "import numpy as np\n",
    "def make_detail_graph(variant, summary_column):\n",
    "    df = pd.read_csv(\"prevalence_summary.tsv\", sep='\\t')\n",
    "    if variant in list(df)[3:]:\n",
    "        df[\"prevalence\"] = df[variant].str.split(\" \").str[0].astype(float)\n",
    "        #df[\"prevalence_percent\"] = df[\"prevalence\"]*100+1\n",
    "        #df[\"prevalence_log\"]=np.log2(df[\"prevalence_percent\"])\n",
    "        max_prevalence = max(df[\"prevalence\"].to_list())\n",
    "        #max_prevalence=7\n",
    "        #max_prevalence=1.0\n",
    "        df[\"sample_size\"] = (\n",
    "            df[variant].str.split(\"/\").str[1].str.replace(\")\", \"\").astype(float)\n",
    "        )\n",
    "        df = df[df[\"sample_size\"] > 0]\n",
    "        # df = df[df[\"Dataset\"] == int(dataset)]\n",
    "\n",
    "        fig = px.scatter_map(\n",
    "            df,\n",
    "            lat=\"Latitude\",\n",
    "            lon=\"Longitude\",\n",
    "            #color=\"prevalence_log\",\n",
    "            color=\"prevalence\",\n",
    "            size=\"sample_size\",\n",
    "            size_max=50,\n",
    "            # color_continuous_scale='cividis',\n",
    "            range_color=(0, max_prevalence), # use this to make the most prevalent mutation \"pop\" as bright yellow\n",
    "            zoom=zoom_level,\n",
    "            hover_name=summary_column,\n",
    "            height=800,\n",
    "            width=800,\n",
    "            hover_data=[\"sample_size\"],\n",
    "            center={\"lat\": latitude, \"lon\": longitude},\n",
    "            title=variant+'_'+summary_column+'_0-'+str(max_prevalence),\n",
    "        )\n",
    "    fig.update_layout(\n",
    "        mapbox=dict(\n",
    "            layers=[\n",
    "                dict(\n",
    "                    sourcetype=\"geojson\",\n",
    "                    source={\"type\": \"FeatureCollection\", \"features\": []},\n",
    "                    type=\"fill\",\n",
    "                    color=\"blue\",\n",
    "                    below=\"traces\",\n",
    "                    visible=False  # Hide the layer by default\n",
    "                    )\n",
    "    ]\n",
    "    )\n",
    "    )\n",
    "    return fig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7e413adb-868d-4781-b77b-23b142e210de",
   "metadata": {},
   "outputs": [],
   "source": [
    "# RUN\n",
    "variants_of_interest=open(\"prevalence_summary.tsv\").readline().split('\\t')[3:]\n",
    "variant = widgets.Dropdown(\n",
    "    options=variants_of_interest,\n",
    "    description='variant:',\n",
    "    disabled=False,\n",
    ")\n",
    "display(variant)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a903a821",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig=make_detail_graph(variant.value, summary_column)\n",
    "fig.show()\n",
    "title_string=f'{variant.value}_{summary_column}'\n",
    "#fig.write_image(title_string+'.svg')\n",
    "#fig.write_image(title_string+'.png')\n",
    "#fig.write_html(title_string+'.html')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f26ed6d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Google AI suggested version (output below) - this version is required if you want to hide rivers, minor cities,\n",
    "# major roads, district boundaries, etc. from the map - but this comes at the expense of loss of country labels.\n",
    "\n",
    "# This code is still not fully implemented. If we went this route, we would need a box for user to manually add\n",
    "# country names and GPS coordinates for each. A full implementation would also add a scale bar, country zoom dropdown\n",
    "# menu, and colored dots like the ones above\n",
    "import plotly.graph_objects as go\n",
    "\n",
    "# Sample data - currently set to some arbitrary points in the US, easily reconfigurable\n",
    "data = {'lat': [40.7128, 34.0522, 41.8781], 'lon': [-74.0060, -118.2437, -87.6298]}\n",
    "\n",
    "fig = go.Figure(go.Scattergeo(\n",
    "    lat=data['lat'],\n",
    "    lon=data['lon'],\n",
    "    mode='markers',\n",
    "    marker=dict(size=10)\n",
    "))\n",
    "\n",
    "fig.update_layout(\n",
    "    geo=dict(\n",
    "        scope='africa',\n",
    "        showland=True,\n",
    "        landcolor='rgb(217, 217, 217)',\n",
    "        showcountries=True,\n",
    "        countrycolor='rgb(255, 255, 255)',\n",
    "        showcoastlines=True,\n",
    "        coastlinecolor='rgb(128, 128, 128)',\n",
    "        showframe=False,  # Remove the outer frame\n",
    "        showlakes=True,\n",
    "        lakecolor='rgb(173, 216, 230)',\n",
    "        showocean=True,\n",
    "        oceancolor='rgb(128, 177, 211)',\n",
    "        projection_type='mercator',  # Use a natural earth projection\n",
    "    ),\n",
    "    margin=dict(l=0, r=0, t=0, b=0)  # Remove margins\n",
    ")\n",
    "\n",
    "fig.update_layout(\n",
    "    autosize=True,\n",
    "    height=800,\n",
    "    geo=dict(\n",
    "        center=dict(\n",
    "            lat=latitude,\n",
    "            lon=longitude\n",
    "        ),\n",
    "        scope='africa',\n",
    "        projection_scale=6.3\n",
    "    )\n",
    ")\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8caeb081-a43f-4c03-be2d-ee8253a63efb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# RUN\n",
    "# This code generates a bar chart for your mutations of interest\n",
    "def get_prevalences(prevalence_file, mutation_list):\n",
    "\tprevalence_dict={'mutations':[], 'prevalences':[], 'gene_name':[]}\n",
    "\th_dict={}\n",
    "\tfor line_number, line in enumerate(open(prevalence_file)):\n",
    "\t\tline=line.strip().split('\\t')\n",
    "\t\tif line_number==0:\n",
    "\t\t\tfor column_number, column in enumerate(line):\n",
    "\t\t\t\th_dict[column]=column_number\n",
    "\t\telif line[0]=='overall':\n",
    "\t\t\tfor mutation in mutation_list:\n",
    "\t\t\t\tif mutation in h_dict:\n",
    "\t\t\t\t\tcolumn=h_dict[mutation]\n",
    "\t\t\t\t\tprevalence=float(line[column].split(' ')[0])\n",
    "\t\t\t\t\tprevalence_dict['mutations'].append(mutation)\n",
    "\t\t\t\t\tprevalence_dict['prevalences'].append(prevalence)\n",
    "\t\t\t\t\tprevalence_dict['gene_name'].append(mutation.split('-')[0])\n",
    "\t\t\t\telse:\n",
    "\t\t\t\t\tprint(mutation, f'is not found in {prevalence_file}!!')\n",
    "\treturn prevalence_dict\n",
    "\n",
    "def make_bar_chart(prevalence_dict, graph_type=\"bar chart of variants_of_interest\", y_max=1.0):\n",
    "\timport pandas as pd\n",
    "\timport plotly.express as px\n",
    "\tdf = pd.DataFrame(prevalence_dict)\n",
    "\tfig1=px.bar(df, x='mutations', y='prevalences', color='gene_name', text_auto=False)\n",
    "\tfig1['layout']['xaxis']['title']=graph_type\n",
    "\tfig1['layout']['yaxis']['title']='overall prevalence'\n",
    "\tfig1.update_layout(yaxis=dict(range=[0, y_max]))\n",
    "\tfig1.write_html(f'{graph_type}.html')\n",
    "\tfig1.write_image(f'{graph_type}.svg')\n",
    "\tfig1.write_image(f'{graph_type}.png')\n",
    "\treturn fig1\n",
    "chart_name='drug resistance mutations'\n",
    "y_max=1.0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3fa7e4c4-3f46-4f5d-85f7-904a9349d893",
   "metadata": {},
   "outputs": [],
   "source": [
    "# OPTIONAL USER INPUT\n",
    "# User can optionally reset the variants_of_interest to graph, chart_name, and y_max values (e.g. by commenting in one\n",
    "#of the blocks of lines below), otherwise the values will be pulled from the cells above\n",
    "\n",
    "#variants_of_interest=['crt-Met74Ile', 'crt-Asn75Glu', 'crt-Lys76Thr','crt-Gln271Glu',\n",
    "#'crt-Arg371Ile', 'dhfr-ts-Asn51Ile', 'dhfr-ts-Cys59Arg', 'dhfr-ts-Ser108Asn',\n",
    "#'dhfr-ts-Ile164Leu', 'dhps-Ala437Gly', 'dhps-Lys540Glu', 'dhps-Ala581Gly',\n",
    "#'mdr1-Asn86Tyr', 'mdr1-Tyr184Phe', 'mdr1-Asp650Asn', 'mdr1-Asn652Asp',\n",
    "#'mdr1-Asp1246Tyr']\n",
    "#chart_name='non-k13 mutations'\n",
    "#y_max=1.0\n",
    "\n",
    "#variants_of_interest=['k13-Lys108Glu', 'k13-His136Asn', 'k13-Lys189Asn', 'k13-Lys189Thr',\n",
    "#'k13-Arg255Lys', 'k13-Leu258Met', 'k13-Gln271His', 'k13-Cys469Phe',\n",
    "#'k13-Cys469Tyr', 'k13-Asn490Thr', 'k13-Arg561His', 'k13-Ala578Ser',\n",
    "#'k13-Ala675Val']\n",
    "#chart_name='k13 mutations'\n",
    "#y_max=0.5\n",
    "\n",
    "\n",
    "# NO USER INPUT BELOW\n",
    "prevalence_dict=get_prevalences('prevalence_summary.tsv', variants_of_interest)\n",
    "fig1=make_bar_chart(prevalence_dict, chart_name, y_max)\n",
    "fig1.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "591fee5f-f823-45dd-a2c5-23796b403bf5",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## Filter genotypes / prevalences\n",
    "It is generally a good idea to do some basic noise removal once the genotypes are created. Some suggestions are provided here."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "86988f8e-902f-47c8-b232-8bf5ae3de379",
   "metadata": {},
   "source": [
    "### Filter variants that are always at low WSAF\n",
    "If a variant is only seen at a low frequency within samples, it is a good indication that it could be just noise. Here we will set a number of samples and minimum WSAF threshold to remove such noise.\n",
    "\n",
    "```python\n",
    "num_samples_wsaf = 2\n",
    "min_wsaf = 0.5\n",
    "wsaf_filter = ((freq > min_wsaf).sum()) >= num_samples_wsaf\n",
    "```\n",
    "\n",
    "The above options will keep the variants that are in at > 0.5 WSAF in at least 2 samples."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3623718f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# USER INPUT\n",
    "num_samples_wsaf = 2\n",
    "min_wsaf = 0.5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c7a1f90d",
   "metadata": {},
   "outputs": [],
   "source": [
    "wsaf_filter = ((freq > min_wsaf).sum()) >= num_samples_wsaf\n",
    "print((\"{} of {} variants will remain after the wsaf filter\").format(\n",
    "    wsaf_filter.sum(), freq.shape[1]))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ca37e13-8923-4c74-b87f-875aeccaf940",
   "metadata": {},
   "source": [
    "### Filter variants that are observed with low UMI counts\n",
    "If a variant is only supported by a low number of UMIs across the entire sample set, it is another indication of noise.\n",
    "\n",
    "```python\n",
    "num_samples_umi = 2\n",
    "min_umi = 3\n",
    "umi_filter = ((filtered_mutation_counts >= min_umi).sum()) > num_samples_umi\n",
    "```\n",
    "\n",
    "The above options will keep the variants that are supported by at least 3 UMIs in at least 2 samples."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5d5e53b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# USER INPUT\n",
    "num_samples_umi = 2\n",
    "min_umi = 3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1887f461",
   "metadata": {},
   "outputs": [],
   "source": [
    "# RUN\n",
    "umi_filter = ((filtered_mutation_counts >= min_umi).sum()) > num_samples_umi\n",
    "print((\"{} of {} variants will remain after the UMI filter\").format(\n",
    "    umi_filter.sum(), freq.shape[1]))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0ba2b6d0-9879-42b2-a8f7-45bada81ab73",
   "metadata": {},
   "source": [
    "### Keep variants that were targeted\n",
    "In most projects there are a number of variants that we would like to report, even if they are not seen in the sample set. We would like to stop those variants from being removed by the above filters."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7e39407b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# RUN\n",
    "targ = freq.columns.get_level_values(\"Targeted\") == \"Yes\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c381acc7-7bd5-4530-9a74-f7c50856a79c",
   "metadata": {},
   "source": [
    "### Combine filters\n",
    "Keep the variants that are either targeted or passing filters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "183dba3b",
   "metadata": {},
   "outputs": [],
   "source": [
    "variant_mask = targ | (wsaf_filter & umi_filter)\n",
    "print((\"{} variants will remain in the final call set.\\n\"\n",
    "       \"{} variants were targeted and will be kept; and {} will be removed by \"\n",
    "       \"the combined UMI and WSAF filters.\").format(\n",
    "    variant_mask.sum(), targ.sum(), (wsaf_filter & umi_filter).sum()))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7e110236-0635-4bf1-ba9c-f31a314db072",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## Filter data tables with the combined filters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0038335c",
   "metadata": {},
   "outputs": [],
   "source": [
    "filtered_genotypes = genotypes.loc[:, variant_mask]\n",
    "filtered_genotypes.to_csv(os.path.join(wdir, \"final_filtered_genotypes.csv\"))\n",
    "filtered_genotypes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4586302f",
   "metadata": {},
   "outputs": [],
   "source": [
    "filtered_prevalences = prevalences.loc[:, variant_mask]\n",
    "filtered_prevalences.to_csv(os.path.join(wdir, \"final_filtered_prevalences_input_table.csv\"))\n",
    "filtered_prevalences.head()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
